<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title> </title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <link rel="stylesheet" href="css/bb.css" />
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title"> </h1>
</header>
<hr />
<p><strong>Arthur Zwaenepoel</strong> (Gent °1995)</p>
<p><img src="img/auv1.jpg" style="width:100.0%" /></p>
<p>I am currently a PhD student in the group of professor <a href="https://www.vandepeerlab.org/">Yves Van de Peer</a>, conducting research on the evolutionary importance of polyploidy. In practice I have mainly worked on statistical phylogenetic methods for the inference of genome duplications from comparative genomic data and modeling of gene family evolution, where I have come to prefer a Bayesian approach towards statistics. More broadly, my main research interests are (theoretical) evolutionary genetics and probabilistic modeling in biology. Besides my aspirations in science, I take a strong interest in philosophy, music and literature. I enjoy computer programming, playing music, long(ish)-distance running and traveling by bicycle.</p>
<p><a href="#publications">publications</a> | <a href="#code">code</a> | <a href="#other">other</a></p>
<hr />
<h2 id="publications">Publications</h2>
<p><a href="https://academic.oup.com/mbe/article/37/9/2734/5828230?guestAccessKey=1c2319f7-f8e5-4b83-878a-4c5ca9a0d35f">Model-Based Detection of Whole-Genome Duplications in a Phylogeny</a><br />
<strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;<br />
2020 – <em>Molecular Biology and Evolution</em>, 37 (9), 2734–2746</p>
<p>The hornwort genome and early land plant evolution<br />
Zhang, Jian; Fu, Xin-Xing; Li, Rui-Qi; Zhao, Xiang; Liu, Yang; Li, Ming-He; <strong>Zwaenepoel, Arthur</strong>; Ma, Hong; Goffinet, Bernard; Guan, Yan-Long; <em>et al.</em><br />
2020 – <em>Nature plants</em> 6 (2), 107-118</p>
<p>Multi-faceted analysis provides little evidence for recurrent whole-genome duplications during hexapod evolution<br />
Roelofs, Dick; <strong>Zwaenepoel, Arthur</strong>; Sistermans, Tom; Nap, Joey; Kampfraath, Andries A; Van de Peer, Yves; Ellers, Jacintha; Kraaijeveld, Ken;<br />
2020 – <em>BMC Biology</em> 18 (1), 1-13</p>
<p>Inference of ancient whole-genome duplications and the evolution of gene duplication and loss rates<br />
<strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;<br />
2019 – <em>Molecular biology and evolution</em> 36 (7), 1384-1404</p>
<p>Finding evidence for whole genome duplications: a reappraisal<br />
<strong>Zwaenepoel, Arthur</strong>; Li, Zhen; Lohaus, Rolf; Van de Peer, Yves;<br />
2019 – <em>Molecular plant</em> 12 (2), 133-136</p>
<p>wgd—simple command line tools for the analysis of ancient whole-genome duplications<br />
<strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;<br />
2019 – <em>Bioinformatics</em> 35 (12), 2153-2155</p>
<p>MorphDB: prioritizing genes for specialized metabolism pathways and gene ontology categories in plants<br />
<strong>Zwaenepoel, Arthur</strong>; Diels, Tim; Amar, David; Van Parys, Thomas; Shamir, Ron; Van de Peer, Yves; Tzfadia, Oren;<br />
2018 – <em>Frontiers in plant science</em> 9 (nan), 352</p>
<hr />
<h2 id="code">Code</h2>
<p><a href="https://github.com/arzwa/Whale.jl">Whale</a>: Flexible Bayesian gene tree reconciliation using amalgamated likelihood estimation and probabilistic programming.</p>
<p><a href="https://github.com/arzwa/Beluga.jl">Beluga</a>: Reversible-jump MCMC for model-based inference of ancient WGD events using gene count data.</p>
<p><a href="https://github.com/arzwa/wgd">wgd</a>: simple command line tools for estimating <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mstyle mathvariant="normal"><mi>S</mi></mstyle></msub><annotation encoding="application/x-tex">K_\mathrm{S}</annotation></semantics></math> distributions, synteny analyses and visualization of these things to unveil ancient WGD events.</p>
<p>… and more on <a href="https://github.com/arzwa/wgd">github</a></p>
<hr />
<h2 id="other">Other</h2>
<p><a href="https://bruinebeer.bandcamp.com/">bruine beer</a>, featured on bandcamp daily’s ‘Best New Ambient Music’ <a href="https://daily.bandcamp.com/best-ambient/best-new-ambient-june-2018">here</a>. Also see <a href="https://www.vuilbak.in">vuilbak recordlabel</a>, also <a href="https://vuilbak.github.io/">here</a>.</p>
<p>A particularly nice <a href="img/ikenrik.jpg">picture of my cat and me</a>.</p>
<p><a href="https://www.discogs.com/user/dingelam">discogs</a>, <a href="https://bandcamp.com/bruinebeer">bandcamp</a></p>
<p>Other biographic trivia:</p>
<ul>
<li>Vegetarian, quasi-vegan</li>
<li>Fairly experienced cook</li>
<li>No driver’s license, strongly opposed to all cars</li>
<li>Favorite programming language: <code>julia</code></li>
<li>Favorite software tools: <code>pandoc</code>, <code>vim</code>, <code>cmus</code></li>
<li>Favorite authors (literature): W. F. Hermans, Hugo Claus, Vladimir Nabokov, Milan Kundera</li>
<li>Cat person, dogs are OK</li>
</ul>
<hr />
<p><a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png" /></a>.</p>
</body>
</html>
