<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="stylesheet" href="/libs/katex/katex.min.css">
     
   <link rel="stylesheet" href="/libs/highlight/github.min.css">
   
  <link rel="stylesheet" href="/css/franklin.css">
  <link rel="stylesheet" href="/css/basic.css">
   <title>ʕ •㉨• ʔ</title>  
</head>
<body>
  <header>
	  <!--div class="blog-name"><a href="/">bb</a></div--->
  </header>

<!-- Content appended here -->
<div class="franklin-content"><p><strong>Arthur Zwaenepoel</strong>  &#40;Gent °1995&#41;. I am currently a PhD student in the group of professor Yves Van de Peer, conducting research on the evolutionary importance of polyploidy. In practice I have mainly worked on statistical phylogenetic methods for the inference of genome duplications from comparative genomic data and modeling of gene family evolution, where I have come to prefer a Bayesian approach towards statistics.  More broadly, my main research interests are &#40;theoretical&#41; evolutionary genetics and probabilistic modeling in biology. I also take a strong interest in philosophy and mathematics, but claim little expertise there. I enjoy music, literature, computer programming, playing music, long&#40;ish&#41;-distance running and traveling by bicycle.</p>
<p><a href="#papers">papers</a> | <a href="#code">code</a> | <a href="#notes">notes</a> | <a href="/pages/photos">photos</a> | <a href="#other">other</a> | <a href="/phylocourse">phylogenetics-course</a></p>
<p><img src="/assets/img/schapen.jpg" alt="" /></p>
<h2 id="papers"><a href="#papers" class="header-anchor">Papers</a></h2>
<p><em>Preprints</em></p>
<p><a href="https://doi.org/10.1101/2021.03.18.435925">A two-type branching process model of gene family evolution</a>. <strong>Arthur Zwaenepoel</strong>, Yves Van de Peer bioRxiv 2021.03.18.435925; doi: https://doi.org/10.1101/2021.03.18.435925</p>
<p><em>Published</em></p>
<p><a href="https://academic.oup.com/mbe/article/37/9/2734/5828230?guestAccessKey&#61;1c2319f7-f8e5-4b83-878a-4c5ca9a0d35f">Model-Based Detection of Whole-Genome Duplications in a Phylogeny.</a>   <strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;   2020 – <em>Molecular Biology and Evolution</em>, 37 &#40;9&#41;, 2734–2746</p>
<p><a href="https://www.nature.com/articles/s41477-019-0588-4">The hornwort genome and early land plant evolution.</a> Zhang, Jian; Fu, Xin-Xing; Li, Rui-Qi; Zhao, Xiang; Liu, Yang; Li, Ming-He; <strong>Zwaenepoel, Arthur</strong>; Ma, Hong; Goffinet, Bernard; Guan, Yan-Long; <em>et al.</em>    2020 – <em>Nature plants</em> 6 &#40;2&#41;, 107-118</p>
<p><a href="https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-020-00789-1">Multi-faceted analysis provides little evidence for recurrent whole-genome duplications during hexapod evolution.</a> Roelofs, Dick; <strong>Zwaenepoel, Arthur</strong>; Sistermans, Tom; Nap, Joey; Kampfraath, Andries A; Van de Peer, Yves; Ellers, Jacintha; Kraaijeveld, Ken;    2020 – <em>BMC Biology</em> 18 &#40;1&#41;, 1-13</p>
<p><a href="https://academic.oup.com/mbe/article/36/7/1384/5475503">Inference of ancient whole-genome duplications and the evolution of gene duplication and loss rates.</a> <strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;    2019 – <em>Molecular biology and evolution</em> 36 &#40;7&#41;, 1384-1404</p>
<p>Finding evidence for whole genome duplications: a reappraisal.   <strong>Zwaenepoel, Arthur</strong>; Li, Zhen; Lohaus, Rolf; Van de Peer, Yves;    2019 – <em>Molecular plant</em> 12 &#40;2&#41;, 133-136</p>
<p>wgd—simple command line tools for the analysis of ancient whole-genome duplications. <strong>Zwaenepoel, Arthur</strong>; Van de Peer, Yves;    2019 – <em>Bioinformatics</em> 35 &#40;12&#41;, 2153-2155</p>
<p>MorphDB: prioritizing genes for specialized metabolism pathways and gene ontology categories in plants   <strong>Zwaenepoel, Arthur</strong>; Diels, Tim; Amar, David; Van Parys, Thomas; Shamir, Ron; Van de Peer, Yves; Tzfadia, Oren;    2018 – <em>Frontiers in plant science</em> 9, 352</p>
<h2 id="code"><a href="#code" class="header-anchor">Code</a></h2>
<p>All my code that is to some extent usable can be found on <a href="https://github.com/arzwa/wgd">github</a>.</p>
<p><a href="https://github.com/arzwa/Whale.jl">Whale</a>: Flexible Bayesian gene tree reconciliation using amalgamated likelihood estimation and probabilistic programming.</p>
<p><a href="https://github.com/arzwa/DeadBird.jl/">DeadBird</a>: Statistical analysis of comparative genomic data using phylogenetic birth-death process models.  Still in a quite experimental stage.</p>
<p><a href="https://github.com/arzwa/TwoTypeDLModel">TwoTypeDLModel</a>: Inference for a two type continuous-time branching process model of gene family evolution. I did not look for a clever name for this one – it&#39;s quite specialized on one thing anyway.</p>
<p><a href="https://github.com/arzwa/Beluga.jl">Beluga</a>: Reversible-jump MCMC for model-based inference of ancient WGD events using gene count data.</p>
<p><a href="https://github.com/arzwa/wgd">wgd</a>: simple command line tools for estimating \(K_\mathrm{S}\) distributions, synteny analyses and visualization of these things to unveil ancient WGD events.</p>
<h2 id="notes"><a href="#notes" class="header-anchor">Notes</a></h2>
<p><a href="posts/2021-01-01-fgm">Fisher&#39;s geometric model and the cost of complexity</a></p>
<h2 id="other"><a href="#other" class="header-anchor">Other</a></h2>
<p><a href="https://bruinebeer.bandcamp.com/">bruine beer</a>, featured on bandcamp daily&#39;s &#39;Best New Ambient Music&#39; <a href="https://daily.bandcamp.com/best-ambient/best-new-ambient-june-2018">here</a>. Also see <a href="https://www.vuilbak.in">vuilbak recordlabel</a>, also <a href="https://vuilbak.github.io/">here</a>.</p>
<p>A particularly nice <a href="/assets/img/ikenrik.jpg">picture of my cat and me</a>.</p>
<p><a href="https://www.discogs.com/user/dingelam">discogs</a>, <a href="https://bandcamp.com/bruinebeer">bandcamp</a></p>
<p>Other biographic trivia:</p>
<ul>
<li><p>Vegetarian, quasi-vegan</p>
</li>
<li><p>Fairly experienced cook</p>
</li>
<li><p>No driver&#39;s license, strongly opposed to all cars</p>
</li>
<li><p>Favorite programming language: <code>julia</code></p>
</li>
<li><p>Favorite software tools: <code>pandoc</code>, <code>vim</code>, <code>cmus</code></p>
</li>
<li><p>Favorite authors &#40;literature&#41;: W. F. Hermans, Hugo Claus, Vladimir Nabokov, Milan Kundera</p>
</li>
<li><p>Cat person, dogs are OK</p>
</li>
</ul>
<div class="page-foot">
  <div class="copyright">
    &copy; Arthur Zwaenepoel. Last modified: March 30, 2021. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>
</div><!-- CONTENT ENDS HERE -->
    
        <script src="/libs/katex/katex.min.js"></script>
<script src="/libs/katex/auto-render.min.js"></script>
<script>renderMathInElement(document.body)</script>

    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
  </body>
</html>
