<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="stylesheet" href="/libs/katex/katex.min.css">
     
   <link rel="stylesheet" href="/libs/highlight/github.min.css">
   
  <link rel="stylesheet" href="/css/franklin.css">
  <link rel="stylesheet" href="/css/basic.css">
   <title>Ks distributions with wgd</title>  
</head>
<body>
  <header>
	  <!--div class="blog-name"><a href="/">bb</a></div--->
  </header>

<!-- Content appended here -->
<div class="franklin-content"><h1 id="computing_ks_distributions_using_wgd"><a href="#computing_ks_distributions_using_wgd" class="header-anchor">Computing \(\ks\) distributions using <code>wgd</code></a></h1>
<p>Already back in 2017, <a href="https://github.com/arzwa/wgd">I wrote a python library</a> for computing so-called \(\ks\)<sup id="fnref:ks"><a href="#fndef:ks" class="fnref">[1]</a></sup> distributions, which can serve to detect ancient whole-genome duplications from genomic data. The goal was to provide an easy and robust pipeline to do exactly that, and it seems some people are in fact using it. It is not very actively maintained &#40;and I wish I&#39;d find the time to finish a complete rewrite, register the package and rewrite documentation&#41; but it does the job nicely and efficiently. Here I&#39;ll show how I generally use the library in my research.</p>
<p>First, installation. I have used wgd with Python3.5&#43; &#40;3.5, 3.6, 3.7 and 3.8&#41;. I generally recommend to use a virtual environment for wgd, that way you won&#39;t have dependency issues or clashes. The following should suffice to  install the wgd package in a Ubuntu-ish Linux environment.</p>
<pre><code class="language-bash">sudo apt-get install python3-pip
pip3 install virtualenv --user
virtualenv venv -p python3
source venv/bin/activate
git clone https://github.com/arzwa/wgd.git
cd wgd
pip3 install .</code></pre>
<p>If all went well, you should try running <code>wgd</code></p>
<pre><code class="language-md">&#40;venv&#41;  &#36; wgd                                                                                                  
Usage: wgd &#91;OPTIONS&#93; COMMAND &#91;ARGS&#93;...

  Welcome to the wgd command line interface&#33;

                         _______
                         \  ___ &#96;&#39;.
         _     _ .--./&#41;   &#39; |--.\  \
   /\    \\   ///.&#39;&#39;\\    | |    \  &#39;
   &#96;\\  //\\ //| |  | |   | |     |  &#39;
     \&#96;//  \&#39;/  \&#96;-&#39; /    | |     |  |
      \|   |/   /&#40;&quot;&#39;&#96;     | |     &#39; .&#39;
       &#39;        \ &#39;---.   | |___.&#39; /&#39;
                 /&#39;&quot;&quot;&#39;.\ /_______.&#39;/
                ||     ||\_______|/
                \&#39;. __//
                 &#96;&#39;---&#39;
  
  wgd  Copyright &#40;C&#41; 2018 Arthur Zwaenepoel
  This program comes with ABSOLUTELY NO WARRANTY;
  This is free software, and you are welcome to redistribute it
  under certain conditions;

  Contact: arzwa@psb.vib-ugent.be

Options:
  -v, --verbosity &#91;info|debug&#93;  Verbosity level, default &#61; info.
  -l, --logfile TEXT            File to write logs to &#40;optional&#41;
  --version                     Print version number
  -h, --help                    Show this message and exit.

Commands:
  dmd  All-vs.-all diamond blastp &#43; MCL clustering.
  kde  Fit a KDE to a Ks distribution.
  ksd  Ks distribution construction.
  mcl  All-vs.-all blastp &#43; MCL clustering.
  mix  Mixture modeling of Ks distributions.
  pre  Check and optionally rename CDS files Example usage &#40;renaming&#41; wgd...
  syn  Co-linearity analyses.
  viz  Plot histograms/densities &#40;interactively&#41;.
  wf1  Standard workflow whole paranome Ks.
  wf2  Standard workflow one-vs-one ortholog Ks.</code></pre>
<p>Some third-party tools are required to run the example below. Specifically we will use <code>diamond</code>, <code>mcl</code>, <code>mafft</code>, <code>codeml</code> and <code>fasttree</code>.</p>
<h2 id="the_data"><a href="#the_data" class="header-anchor">The data</a></h2>
<p>The basic input data is a bunch of CDS sequences. Here I&#39;ll use the CDS data available for the bladderwort &#40;<em>Utricularia gibba</em>&#41;. To download the data  from PLAZA:</p>
<pre><code class="language-bash">wget ftp://ftp.psb.ugent.be/pub/plaza/plaza_public_dicots_04/Fasta/cds.selected_transcript.ugi.fasta.gz
gunzip *gz
mv cds.*.ugi.fasta ugi.fasta</code></pre>
<p><img src="https://newfs.s3.amazonaws.com/taxon-images-1000s1000/Lentibulariaceae/utricularia-gibba-fl-dcameron.jpg" alt="" /></p>
<p>There is a little tool in <code>wgd</code> to quickly check the input data. Since \(\ks\) is a codon-model based evolutionary distance, <em>it is absolutely crucial that the input data are proper codon-sequences</em>. If you&#39;re paranoid about this being  the case, you can use <code>wgd pre</code> to partition a fasta file in everything that is nicely translatable from start to stop codon and all the rest.</p>
<pre><code class="language-bash">wgd pre ugi.fasta</code></pre>
<p>This will spit a lot to the screen:</p>
<pre><code class="language-julia">&#36; wgd pre ugi.fasta 
2020-03-07 19:42:14: INFO	&#40;0&#41; Checking ugi.fasta
2020-03-07 19:42:14: ERROR	Translation error &#40;First codon &#39;CCG&#39; is not a start codon&#41; in sequence UGI.ctg09795.24681.1
2020-03-07 19:42:14: ERROR	Translation error &#40;First codon &#39;AAA&#39; is not a start codon&#41; in sequence UGI.ctg09872.24683.1
2020-03-07 19:42:14: ERROR	Translation error &#40;Sequence length 638 is not a multiple of three&#41; in sequence UGI.ctg09892.24686.1
2020-03-07 19:42:14: ERROR	Translation error &#40;First codon &#39;TAG&#39; is not a start codon&#41; in sequence UGI.ctg09909.24689.1
&#91;...&#93;
2020-03-07 19:42:21: INFO	24145/25930 &#40;93.12&#37;&#41; sequences are perfect CDS &#40;in ugi.fasta.pre.good&#41;
2020-03-07 19:42:21: INFO	1785/25930 &#40;6.88&#37;&#41; sequences are not perfect CDS &#40;in ugi.fasta.pre.bad&#41;</code></pre>
<p>About 93&#37; of the input data is textbook CDS sequence. The rest has some issue, and is written to the <code>.bad</code> file. Of course, not all CDS sequences start with a start codon, and if you trust your data, you should definitely not throw those away&#33; &#40;you should, however, not trust any sequence with a length that is no multiple of three&#33;&#41;. Here I&#39;ll continue working with the &#39;proper&#39; CDS data, now in the file <code>ugi.fasta.pre.good</code>.</p>
<p><strong>Note:</strong> <code>wgd pre</code> can also be used to rename your sequences. If you have the kind of fasta files with very long and awkward headers, I would recommend this, as it will make files obtained later clearer. See the <code>--rename</code> and <code>--prefix</code> options.</p>
<h2 id="obtaining_the_paranome"><a href="#obtaining_the_paranome" class="header-anchor">Obtaining the paranome</a></h2>
<p>The first step is to obtain the <strong>paranome</strong>, or the collection of all paralogous genes. This is esentially a big graph, where the nodes are all genes in the genome, and edges represent homology &#40;paralogy&#41; relationships. Of course, in the absence of different genomes to compare to, homology is no well-defined concept, as we may as well assume all genes trace back to a common ancestor  and are thus homologous. The goal of paranome inference is of course not to lump everything together, but simply to <em>cluster the genome in reasonably fine-grained  paralogous gene families</em>. Ideally, we&#39;d like to obtain paralogous gene families with a most recent common ancestor &#40;MRCA&#41; that is within the time frame where \(\ks\) can be reliably estimated, providing us a clue as to what reasonably  fine-grained could mean.</p>
<p>In general, common gene family clustering methods using Markov graph clustering  &#40;MCL&#41; work well for this task. In wgd an approach based on all-vs.-all protein similarity searches and MCL clustering is implemented. Both <code>blastp</code>, and the  much faster <code>diamond</code> are supported. The following will run <code>diamond</code> &#43; <code>mcl</code> to obtain the paranome:</p>
<pre><code class="language-julia">wgd dmd ugi.fasta.pre.good</code></pre>
<p>If one wishes to use the full data &#40;not only the textbook-CDS sequences&#41;, the <code>--nostrictcds</code> and <code>--ignorestop</code> options can be used. Other parameters of interest are the \(e\)-value threshold used to construct the sequence-similarity graph and the inflation factor for MCL, governing the coarseness of the inferred clusters. To see all options for <code>wgd dmd</code> run <code>wgd dmd --help</code>.</p>
<p>If we started from an empty directory, by now we should have obtained the  following &#40;using the non-default <code>tree</code> command in Linux&#41;:</p>
<pre><code class="language-bash">&#36; tree   
.
├── ugi.fasta
├── ugi.fasta.pre.bad
├── ugi.fasta.pre.good
└── wgd_dmd
    └── ugi.fasta.mcl</code></pre>
<p>The paranome consists of how many families?</p>
<pre><code class="language-julia">&#36; wc -l wgd_dmd/ugi.fasta.pre.good.mcl
3086 wgd_dmd/ugi.fasta.mcl</code></pre>
<h2 id="the_paranome_ks_distribution"><a href="#the_paranome_ks_distribution" class="header-anchor">The paranome \(\ks\) distribution</a></h2>
<p>To compute the \(\ks\) distribution, we&#39;ll perform multiple sequence alignment with MAFFT, maximum likelihood estimation of codon substitution model parameters using <code>codeml</code> and do approximate phylogenetic tree inference using FastTree. </p>
<p>Since this takes quite a bit of time &#40;codeml performs a rather expensive  numerical optimization step&#41;, one would usually run this on a computing  cluster on a couple of CPUs. This data set is still feasible to process  on a laptop with 4 cores if we filter out some big families.</p>
<p>If you have the required tools installed, you should be able to run the following:</p>
<pre><code class="language-julia">wgd ksd ./wgd_dmd/ugi.fasta.mcl ./ugi.fasta.pre.good -mp 1000</code></pre>
<p>&#40;where the <code>mp 1000</code> option filters out all families with more than a thousand pairs of sequences, futrher decrease this number if you want to quickly test the whole thing, since with <code>mp 1000</code> it will still take about two hours&#41;. This will run the analysis in parallel on four CPU cores. It starts with the biggest families &#40;which take more time&#41;. We can monitor the analysis in the terminal, with output that looks like this.</p>
<pre><code class="language-julia">&#36;  wgd ksd ./wgd_dmd/ugi.fasta.mcl ./ugi.fasta.pre.good  
&#91;...&#93;
2020-03-07 19:01:43: INFO	Performing analysis on gene family GF_003081
2020-03-07 19:01:44: INFO	Performing analysis on gene family GF_003082
2020-03-07 19:01:44: INFO	Performing analysis on gene family GF_003083
2020-03-07 19:01:44: INFO	Performing analysis on gene family GF_003084
2020-03-07 19:01:44: INFO	Performing analysis on gene family GF_003085
2020-03-07 19:01:44: INFO	Performing analysis on gene family GF_003086
2020-03-07 19:01:45: INFO	Analysis done
2020-03-07 19:01:45: INFO	Making results data frame
2020-03-07 19:02:17: INFO	Removing tmp directory
2020-03-07 19:02:18: INFO	Computing weights, outlier cut-off at Ks &gt; 5
2020-03-07 19:02:19: INFO	Generating plots
2020-03-07 19:02:19: INFO	Will plot **node-weighted** histograms
2020-03-07 19:02:20: INFO	Done</code></pre>
<p>The generated output is in the <code>wgd_ksd</code> directory</p>
<pre><code class="language-julia">&#36; tree wgd_ksd
wgd_ksd
├── ugi.fasta.pre.good.ks.svg
└── ugi.fasta.pre.good.ks.tsv</code></pre>
<p>The main output is the <code>.tsv</code> file, which contains all the results computed in the <code>wgd ksd</code> pipeline:</p>
<pre><code class="language-julia">&#36; head wgd_ksd/ugi.fasta.pre.good.ks.tsv 
	AlignmentCoverage	AlignmentIdentity	AlignmentLength	AlignmentLengthStripped	Distance	Family	Ka	Ks	Node	Omega	Paralog1	Paralog2	WeightOutliersIncluded	WeightOutliersExcluded
UGI.Scf00428.21433.1__UGI.Scf00767.18891.1	0.20142	0.71127	2115.0	426.0	0.51238	GF_002963	0.3227	0.6327	2.0	0.51UGI.Scf00428.21433.1	UGI.Scf00767.18891.1	1.0	1.0
UGI.Scf00223.11926.1__UGI.Scf00720.18573.1	0.31967	0.74359	732.0	234.0	0.48336	GF_002829	0.2658	0.5949	2.0	0.4467	UGI.Scf00223.11926.1	UGI.Scf00720.18573.1	1.0	1.0
UGI.Scf00013.1922.1__UGI.Scf00108.8260.1	0.84677	0.87302	372.0	315.0	0.12595	GF_002055	0.0797	0.3366	2.0	0.2369	UGI.Scf00013.1922.1	UGI.Scf00108.8260.1	1.0	1.0
UGI.Scf00015.2084.1__UGI.Scf00046.4873.1	0.30266	0.78667	1239.0	375.0	0.32056	GF_000243	0.2379	1.5895	10.0	0.1497	UGI.Scf00015.2084.1	UGI.Scf00046.4873.1	1.0	1.0
UGI.Scf00015.2084.1__UGI.Scf00051.5189.1	0.24455	0.56766	1239.0	303.0	0.99481	GF_000243	0.6446	118.4869	11.00.0054	UGI.Scf00051.5189.1	UGI.Scf00015.2084.1	0.5	0.0
UGI.Scf00046.4873.1__UGI.Scf00051.5189.1	0.23002	0.5193	1239.0	285.0	1.07317	GF_000243	4.2678	71.3586	11.0	0.0598	UGI.Scf00051.5189.1	UGI.Scf00046.4873.1	0.5	0.0
UGI.Scf00051.5189.1__UGI.Scf00090.7423.1	0.23729	0.53061	1239.0	294.0	1.09756	GF_000243	13.2138	38.8342	12.0	0.3403	UGI.Scf00090.7423.1	UGI.Scf00051.5189.1	0.33333	0.0
UGI.Scf00015.2084.1__UGI.Scf00090.7423.1	0.3293	0.625	1239.0	408.0	0.61336	GF_000243	16.5071	17.2326	12.0	0.9579	UGI.Scf00090.7423.1	UGI.Scf00015.2084.1	0.33333	0.0
UGI.Scf00046.4873.1__UGI.Scf00090.7423.1	0.33898	0.62381	1239.0	420.0	0.69171	GF_000243	3.9448	34.5095	12.0	0.1143	UGI.Scf00090.7423.1	UGI.Scf00046.4873.1	0.33333	0.0</code></pre>
<p>The default plot outputted by <code>wgd</code> &#40;in <code>wgd_ksd/*.svg</code>&#41; is rather ugly, but gives a clear overview of the parameter estimates for all gene duplication events, being \(\ks, K_\mathrm{A}\) and \(\omega\) &#40;i.e. the sysnonymous distance, nonsynonyous distance and nonsynonymous to synonymous substitution rate ratio respectively&#41;.</p>
<p><img src="/assets/wgd/ugi-ks.svg" alt="" /></p>
<p>From this distribution it&#39;s quite clear there has been some stuff going on  with this genome.  </p>
<h2 id="computing_one-to-one_ortholog_ks_distributions"><a href="#computing_one-to-one_ortholog_ks_distributions" class="header-anchor">Computing one-to-one ortholog \(\ks\) distributions</a></h2>
<p>People often also estimate one-to-one ortholog \(\ks\) distributions to obtain an idea of the relative timing of WGDs and species divergences, divergence times or relative substitution rates. Let us compare <em>Utricularia</em> to a close relative <em>Erythranthe</em>. First get and clean the data</p>
<pre><code class="language-bash">wget ftp://ftp.psb.ugent.be/pub/plaza/plaza_public_dicots_04/Fasta/cds.selected_transcript.egut.fasta.gz
gunzip *gz
mv cds.*.egut.fasta egu.fasta
wgd pre egu.fasta</code></pre>
<p>To obtain one-to-one orthologs for between species comparisons we can again  use <code>diamond</code> with <code>wgd dmd</code>:</p>
<pre><code class="language-bash">wgd dmd egu.fasta.pre.good ugi.fasta.pre.good</code></pre>
<pre><code class="language-julia">&#36; wgd dmd egu.fasta.pre.good ugi.fasta.pre.good
2020-03-07 20:06:51: WARNING	dir wgd_dmd exists&#33;
2020-03-07 20:06:58: WARNING	dir wgd_dmd exists&#33;
2020-03-07 20:07:04: INFO	Multiple CDS files: will compute RBH orthologs
2020-03-07 20:07:04: INFO	egu.fasta.pre.good vs. ugi.fasta.pre.good</code></pre>
<p>The ouput is, as expected, a bunch of gene pairs. These are <em>reciprocal best hits</em> &#40;RBHs&#41;. In other words, two genes \(A\) and \(B\)  are considered one-to-one orthologs  if \(B\) has the highest sequence similarity to \(A\) among the total set of genes &#40;excluding \(A\) itself of course&#41; and vice versa. Have a look at the file to  convince yourself:</p>
<pre><code class="language-julia">&#36; head wgd_dmd/egu.fasta.pre.good_ugi.fasta.pre.good.rbh
UGI.Scf00140.9494.1	Migut.L01925.1
UGI.Scf00873.19408.1	Migut.D02258.1
UGI.Scf01221.23016.1	Migut.K00468.1
UGI.Scf00487.21516.1	Migut.F00288.1
UGI.Scf00033.3844.1	Migut.G01067.1
UGI.Scf00611.21794.1	Migut.N02364.1
UGI.Scf00611.21793.1	Migut.N02367.1
UGI.Scf00131.9250.1	Migut.F01756.1
UGI.Scf00437.15873.1	Migut.F00666.1
UGI.Scf00015.2103.1	Migut.O00089.1</code></pre>
<p>To compute the one-to-one ortholog \(\ks\) distribution, we can again  use <code>wgd ksd</code> intuitively. For the sake of comutational feasibility  I&#39;ll take a random subset first</p>
<pre><code class="language-julia">shuf wgd_dmd/egu.fasta.pre.good_ugi.fasta.pre.good.rbh wgd_dmd/egu-ugi.2500.rbh</code></pre>
<pre><code class="language-julia">&#36; wgd ksd wgd_dmd/egu-ugi.2500.rbh egu.fasta.pre.good ugi.fasta.pre.good
2020-03-07 20:10:02: INFO	
2020-03-07 20:10:02: INFO	codeml found
2020-03-07 20:10:02: INFO	
2020-03-07 20:10:02: INFO	
2020-03-07 20:10:02: INFO	Translating CDS file
Invalid codon CTN in UGI.ctg34642.25621.1                                                       
Invalid codon TCN in UGI.Scf00004.766.1
100&#37; &#40;50833 of 50833&#41; |######################################################| Elapsed Time: 0:00:14 Time:  0:00:14
2020-03-08 12:40:40: WARNING	There were 56 warnings during translation
2020-03-08 12:40:40: INFO	Started whole paranome Ks analysis
2020-03-08 12:40:40: WARNING	Filtered out the 0 largest gene families because n*&#40;n-1&#41;/2 &gt; &#96;max_pairwise&#96;
2020-03-08 12:40:40: WARNING	If you want to analyse these large families anyhow, please raise the &#96;max_pairwise&#96; parameter. 
2020-03-08 12:40:40: INFO	Started analysis in parallel &#40;n_threads &#61; 4&#41;
2020-03-08 12:40:40: INFO	Performing analysis on gene family GF_000001
2020-03-08 12:40:40: INFO	Performing analysis on gene family GF_000002
&#91;...&#93;
2020-03-08 12:51:18: INFO	Performing analysis on gene family GF_002499
2020-03-08 12:51:18: INFO	Performing analysis on gene family GF_002500
2020-03-08 12:51:19: INFO	Analysis done
2020-03-08 12:51:19: INFO	Making results data frame
2020-03-08 12:51:38: INFO	Removing tmp directory
2020-03-08 12:51:38: INFO	Computing weights, outlier cut-off at Ks &gt; 5
2020-03-08 12:51:38: INFO	Generating plots
2020-03-08 12:51:38: INFO	Will plot **node-weighted** histograms
2020-03-08 12:51:39: INFO	Done</code></pre>
<p>Again the default plots are pretty ugly, but show what we need to see.</p>
<p><img src="/assets/wgd/ugi-egu-ks.svg" alt="" /></p>
<h2 id="comparing_multiple_distributions"><a href="#comparing_multiple_distributions" class="header-anchor">Comparing multiple distributions</a></h2>
<p>Say we want to make a nice plot combining the two distributions above.  One can use the utilities iplemented in <code>wgd viz</code>, however, if you&#39;re somewhat familiar with Python, R, julia, or any other programming  language with nice plotting libraries I would defintely recommend  generating your own plots. Here I show an example of how I would go on  visualizing the two distributions computed above comparatively using  Python.</p>
<p>I use the following packages</p>
<pre><code class="language-python">import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns</code></pre>
<p>Then load the data</p>
<pre><code class="language-python">egu_ugi &#61; pd.read_csv&#40;&quot;./wgd_ksd/egu.fasta.pre.good_ugi.fasta.pre.good.ks.tsv&quot;, sep&#61;&#39;\t&#39;, index_col&#61;0&#41;
ugi &#61; pd.read_csv&#40;&quot;./wgd_ksd/ugi.fasta.pre.good.ks.tsv&quot;, sep&#61;&#39;\t&#39;, index_col&#61;0&#41;</code></pre>
<p>Now I filter the \(\ks\) data and perform node averaging. When computing paranome \(\ks\) distributions <code>wgd ksd</code> infers a phylogenetic tree for each paralogous famiy. Using these trees, we can weigh or average \(\ks\) estimates for ancestral duplication events in families to account for the fact that we have multiple redundant \(\ks\) estimates pertaining to the same divergence &#40;duplication&#41; event. I first filter the data frame to only retain \(\ks\) estimates in a reasonable range &#40;here I  take \(0.01 < \hat{\ks} 5\)&#41;, then I will average \(\ks\) estimates that correspond to the same duplication node in each paralogous family. For the one-to-one ortholog distribution the latter is not necessary, and I&#39;ll just apply the same \(0.01 < \hat{\ks} < 5\) filter:</p>
<pre><code class="language-python">lower, upper &#61; 0.01, 5
ugi &#61; ugi&#91;ugi&#91;&quot;Ks&quot;&#93; &lt; upper&#93;
ugi &#61; ugi&#91;ugi&#91;&quot;Ks&quot;&#93; &gt; lower&#93;
ugi_ks &#61; ugi.groupby&#40;&#91;&quot;Family&quot;, &quot;Node&quot;&#93;&#41;&#91;&quot;Ks&quot;&#93;.apply&#40;np.mean&#41;
egu_ugi_ks &#61; egu_ugi&#91;egu_ugi&#91;&quot;Ks&quot;&#93; &lt; upper&#93;&#91;&quot;Ks&quot;&#93;</code></pre>
<p>Now we can go on plotting:</p>
<pre><code class="language-python">fig, ax &#61; plt.subplots&#40;1, 1, figsize&#61;&#40;6,4&#41;&#41;
ax.hist&#40;ugi_ks, bins&#61;np.linspace&#40;0.0, upper, 50&#41;, 
        color&#61;&quot;teal&quot;, alpha&#61;0.2, 
        label&#61;&quot;&#36;U.\ gibba&#36;&quot;&#41;;
ax.hist&#40;egu_ugi_ks, bins&#61;np.linspace&#40;0.0, upper, 50&#41;, 
        color&#61;&quot;black&quot;, alpha&#61;0.2, 
        label&#61;&quot;&#36;U.\ gibba&#36; vs. &#36;E.\ guttata&#36;&quot;&#41;;
ax.set_xlabel&#40;&quot;&#36;\hat&#123;K_\mathrm&#123;S&#125;&#125;&#36;&quot;, fontsize&#61;14&#41;
ax.set_ylabel&#40;&quot;# of duplications/ortholog pairs&quot;, fontsize&#61;14&#41;
ax.set_xlim&#40;0, upper&#41;
ax.legend&#40;frameon&#61;False&#41;;
sns.despine&#40;offset&#61;5&#41;</code></pre>
<p><img src="/assets/wgd/py-ks.svg" alt="" /></p>
<p>This plot merits some interpretation. Of course, \(\ks\) distributions admit only interpretations in terms of broad-scale patterns, and deducing a detailed genome evolutionary history from a &#40;bunch of&#41; \(\ks\) distribution&#40;s&#41; would be like reading tea leaves. In the case of <em>Utricularia</em> we see strong evidence for a WGD associated with the \(\ks \approx 0.5\) peak, and there is a strong suggestion for a second, somewhat older, WGD associated feature. Both seem to have occurred after divergence from <em>Erythranthe</em> &#40;it least if substitution rates do not differ drmatically among thee lineages&#41;. Note that synteny analyses suggested three WGDs in the recent evolutionary past of <em>Utricularia</em>, and while the \(\ks\) distribution does not decisively indicate absence of a third WGD after the <em>Uricularia</em> – <em>Erythranthe</em> divergence, it does not admit such a conclusion either. We further note the very low number of recent gene duplications, in accord with the small genome size and generaal &#39;genomic reduction&#39; described for this species.</p>
<h2 id="mixture_models"><a href="#mixture_models" class="header-anchor">Mixture models</a></h2>
<p>In order to guide interpretation of a \(\ks\) distribution, one-dimensional Gaussian mixture models &#40;GMMs&#41; can be very helpful. It is tempting however to use GMMs for <em>inference</em> purposes in the context of genome evolution. </p>
<pre><code class="language-bash">wgd mix ./wgd_ksd/ugi.fasta.pre.good.ks.tsv -n 2 8</code></pre>
<pre><code class="language-julia">&#36; wgd mix ./wgd_ksd/ugi.fasta.pre.good.ks.tsv -n 2 8
2020-03-14 11:13:15: INFO	Preparing data frame
2020-03-14 11:13:15: INFO	 .. max_iter &#61; 1000
2020-03-14 11:13:15: INFO	 .. n_init   &#61; 1
2020-03-14 11:13:15: INFO	Method is GMM, interpret best model with caution&#33;
2020-03-14 11:13:15: INFO	Fitting GMM with 2 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 0.595, 0.645, 0.685
2020-03-14 11:13:15: INFO	.. 1.865, 0.089, 0.315
2020-03-14 11:13:15: INFO	Fitting GMM with 3 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 0.126, 0.888, 0.066
2020-03-14 11:13:15: INFO	.. 1.705, 0.122, 0.483
2020-03-14 11:13:15: INFO	.. 0.538, 0.142, 0.451
2020-03-14 11:13:15: INFO	Fitting GMM with 4 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 0.144, 0.891, 0.078
2020-03-14 11:13:15: INFO	.. 2.087, 0.048, 0.313
2020-03-14 11:13:15: INFO	.. 0.480, 0.074, 0.344
2020-03-14 11:13:15: INFO	.. 1.059, 0.065, 0.265
2020-03-14 11:13:15: INFO	Fitting GMM with 5 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 0.186, 0.246, 0.060
2020-03-14 11:13:15: INFO	.. 2.083, 0.048, 0.315
2020-03-14 11:13:15: INFO	.. 1.048, 0.063, 0.272
2020-03-14 11:13:15: INFO	.. 0.053, 0.847, 0.020
2020-03-14 11:13:15: INFO	.. 0.482, 0.062, 0.333
2020-03-14 11:13:15: INFO	Fitting GMM with 6 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 1.450, 0.041, 0.207
2020-03-14 11:13:15: INFO	.. 0.469, 0.055, 0.311
2020-03-14 11:13:15: INFO	.. 0.174, 0.274, 0.064
2020-03-14 11:13:15: INFO	.. 0.883, 0.047, 0.188
2020-03-14 11:13:15: INFO	.. 0.040, 0.771, 0.015
2020-03-14 11:13:15: INFO	.. 2.317, 0.024, 0.216
2020-03-14 11:13:15: INFO	Fitting GMM with 7 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 0.503, 0.038, 0.257
2020-03-14 11:13:15: INFO	.. 2.320, 0.024, 0.216
2020-03-14 11:13:15: INFO	.. 0.125, 0.184, 0.041
2020-03-14 11:13:15: INFO	.. 1.472, 0.036, 0.197
2020-03-14 11:13:15: INFO	.. 0.339, 0.074, 0.091
2020-03-14 11:13:15: INFO	.. 0.032, 0.641, 0.012
2020-03-14 11:13:15: INFO	.. 0.912, 0.037, 0.186
2020-03-14 11:13:15: INFO	Fitting GMM with 8 components
2020-03-14 11:13:15: INFO	Component mean, variance, weight: 
2020-03-14 11:13:15: INFO	.. 2.337, 0.022, 0.210
2020-03-14 11:13:15: INFO	.. 0.544, 0.029, 0.197
2020-03-14 11:13:15: INFO	.. 0.015, 0.330, 0.004
2020-03-14 11:13:15: INFO	.. 0.936, 0.034, 0.186
2020-03-14 11:13:15: INFO	.. 0.186, 0.145, 0.048
2020-03-14 11:13:15: INFO	.. 1.504, 0.034, 0.194
2020-03-14 11:13:15: INFO	.. 0.393, 0.034, 0.140
2020-03-14 11:13:15: INFO	.. 0.070, 0.223, 0.020
2020-03-14 11:13:15: INFO	
2020-03-14 11:13:15: INFO	AIC assessment:
2020-03-14 11:13:15: INFO	min&#40;AIC&#41; &#61; 11011.75 for model 6
2020-03-14 11:13:15: INFO	Relative probabilities compared to model 6:
2020-03-14 11:13:15: INFO	   /                          \
2020-03-14 11:13:15: INFO	   |      &#40;min&#40;AIC&#41; - AICi&#41;/2 |
2020-03-14 11:13:15: INFO	   | p &#61; e                    |
2020-03-14 11:13:15: INFO	   \                          /
2020-03-14 11:13:15: INFO	.. model   1: p &#61; 0.0000
2020-03-14 11:13:15: INFO	.. model   2: p &#61; 0.0000
2020-03-14 11:13:15: INFO	.. model   3: p &#61; 0.0000
2020-03-14 11:13:15: INFO	.. model   4: p &#61; 0.0000
2020-03-14 11:13:15: INFO	.. model   5: p &#61; 0.4952
2020-03-14 11:13:15: INFO	.. model   6: p &#61; 1.0000
2020-03-14 11:13:15: INFO	.. model   7: p &#61; 0.0358
2020-03-14 11:13:15: INFO	
2020-03-14 11:13:15: INFO	
2020-03-14 11:13:15: INFO	Delta BIC assessment: 
2020-03-14 11:13:15: INFO	min&#40;BIC&#41; &#61; 11123.79 for model 5
2020-03-14 11:13:15: INFO	.. model   1: delta&#40;BIC&#41; &#61;   940.79 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	.. model   2: delta&#40;BIC&#41; &#61;   297.33 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	.. model   3: delta&#40;BIC&#41; &#61;    45.97 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	.. model   4: delta&#40;BIC&#41; &#61;    38.33 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	.. model   5: delta&#40;BIC&#41; &#61;     0.00 &#40;0 to  2:   Very weak&#41;
2020-03-14 11:13:15: INFO	.. model   6: delta&#40;BIC&#41; &#61;    18.12 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	.. model   7: delta&#40;BIC&#41; &#61;    44.30 &#40;    &gt;10: Very Strong&#41;
2020-03-14 11:13:15: INFO	
2020-03-14 11:13:15: INFO	Plotting AIC &amp; BIC
2020-03-14 11:13:16: INFO	Plotting mixtures
2020-03-14 11:13:19: INFO	Writing component-wise probabilities to file</code></pre>
<p>The AIC &#40;Akaike information criterion&#41; suggests model 6 &#40;the model with seven components&#41; to be the best fitting model, whereas the BIC &#40;Bayesian information criterion&#41; suggests model 5 &#40;which has 6 components&#41; to be the best model. However, plots of the AIC and BIC for increasing model complexity indicate that around four components we start entering a kind  of plateau region:</p>
<p><img src="/assets/wgd/aic_bic.svg" alt="" /></p>
<p>However, as I already stressed above, relying on model selection techniques to do <em>inference</em> of &#39;the correct model&#39; is a hazardous business. Our main aim should be is to present a reasonable analysis of the complicated \(\ks\) distributions in simpler, unimodal components. </p>
<p><img src="/assets/wgd/gmms.svg" alt="" /></p>
<hr />
<table class="fndef" id="fndef:ks">
    <tr>
        <td class="fndef-backref"><a href="#fnref:ks">[1]</a></td>
        <td class="fndef-content">Here I&#39;ll use \(\ks\) and \(K_\mathrm{A}\) instead of the more commonly used \(d\mathrm{S}\) and \(d\mathrm{N}\). This is not due to personal preference &#40;in fact, I would prefer following Ziheng Yang&#39;s notation&#41;, but simply because this is customary in the WGD related literature.</td>
    </tr>
</table>

<div class="page-foot">
  <div class="copyright">
    &copy; Arthur Zwaenepoel. Last modified: March 30, 2021. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>
</div><!-- CONTENT ENDS HERE -->
    
        <script src="/libs/katex/katex.min.js"></script>
<script src="/libs/katex/auto-render.min.js"></script>
<script>renderMathInElement(document.body)</script>

    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
  </body>
</html>
