<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="stylesheet" href="/libs/katex/katex.min.css">
     
   <link rel="stylesheet" href="/libs/highlight/github.min.css">
   
  <link rel="stylesheet" href="/css/franklin.css">
  <link rel="stylesheet" href="/css/basic.css">
   <title>Distance methods</title>  
</head>
<body>
  <header>
	  <!--div class="blog-name"><a href="/">bb</a></div--->
  </header>

<!-- Content appended here -->
<div class="franklin-content"><div class="franklin-toc"><ol><li><a href="#distance_methods">Distance methods</a><ol><li><a href="#distance_matrix">Distance matrix</a></li><li><a href="#clustering_methods">Clustering methods</a></li><li><a href="#neighbor-joining">Neighbor-joining</a></li><li><a href="#gamma_distances">Gamma distances</a></li></ol></li><li><a href="#maximum_likelihood_phylogeny_inference">Maximum likelihood phylogeny inference</a><ol><li><a href="#analysis_under_the_jukes_cantor_model">Analysis under the Jukes &amp; Cantor model</a></li><li><a href="#gamma_distributed_rates">Gamma distributed rates</a></li><li><a href="#other_common_substitution_models">Other common substitution models</a></li><li><a href="#the_bootstrap">The bootstrap</a></li></ol></li></ol></div>
<h1 id="distance_methods"><a href="#distance_methods" class="header-anchor">Distance methods</a></h1>
<h2 id="distance_matrix"><a href="#distance_matrix" class="header-anchor">Distance matrix</a></h2>
<blockquote>
<p>Question: does this distance matrix make sense to you? Can you spot clades of more closely related species already?</p>
</blockquote>
<p>It does make sense at first sight. We see a couple of clear clusters. Clearly tetrapod animals cluster strongly together with a low pairwise distance &#40;Homo and Xenopus&#41;. We see higher plants also forming a tight cluster &#40;Zamia, Zea, Glycine and Oryza&#41;, and the same holds for fungi and green algae. Palmaria &#40;a red alga&#41; seems to be a bit of an outlier.</p>
<blockquote>
<p>Question: consider the portion of the matrix corresponding to the animals &#40;human, Xenopus, Artemia and Anemonia&#41;, write down the associated part of the distance matrix. Confirm that the values are proper distances. Make a sketch of what you think the phylogeny will look like for these four species based on a quick look at the matrix.</p>
</blockquote>
<table><tr><th align="right"></th><th align="right">Human</th><th align="right"><em>Xenopus</em></th><th align="right"><em>Artemia</em></th><th align="right"><em>Anemonia</em></th></tr><tr><td align="right">Human</td><td align="right">0</td><td align="right">0.049</td><td align="right">0.20</td><td align="right">0.22</td></tr><tr><td align="right"><em>Xenopus</em></td><td align="right">0.049</td><td align="right">0</td><td align="right">0.19</td><td align="right">0.21</td></tr><tr><td align="right"><em>Artemia</em></td><td align="right">0.20</td><td align="right">0.19</td><td align="right">0</td><td align="right">0.23</td></tr><tr><td align="right"><em>Anemonia</em></td><td align="right">0.22</td><td align="right">0.21</td><td align="right">0.23</td><td align="right">0</td></tr></table>
<p>Clearly the values are distances, as the distance of \(a\) to \(b\) \(d(a,b)\) is &#40;1&#41; symmetric &#40;\(d(a,b) = d(b,a)\)&#41;, &#40;2&#41; non-negative \(d(a,b) \le 0\) &#40;with equality only when \(a = b\)&#41; and &#40;3&#41; the distances satisfy the triangle inequality that \(d(a,c) \le d(a, c) + d(b, c)\). As a sketch of the phylogeny, I would group Human and <em>Xenopus</em> first, for they are the closest to each other. Then I would group the resulting clade with <em>Artemia</em>, since both human and <em>Xenopus</em> are closer to <em>Artemia</em> than to <em>Anemonia</em>, and since <em>Artemia</em> is closer to both Human and <em>Xenopus</em> then to <em>Anemonia</em>. Lastly I would group the resulting clade with <em>Anemonia</em> as sister lineage.</p>
<h2 id="clustering_methods"><a href="#clustering_methods" class="header-anchor">Clustering methods</a></h2>
<blockquote>
<p>Question: how can you see in one glance that this is an ultrametric tree? </p>
</blockquote>
<p>The distance from tip to root is identical for all tips &#40;leaves&#41; of the tree.</p>
<p>Note that the evolutionary distance represents the estimated ‘amount’ of evolution, i.e. the expected number of substitutions per site, so the feature that all root-to-tip distances are equal amounts to the assumption that the total ‘amount of evolution’ &#40;i.e. expected number of substitutions&#41; in the evolutionary history of each taxon is the same. Since the actual time &#40;in millions of years say&#41; to the root is the same for all tips of the tree, this is equivalent to the assumption that the average substitution rate is equal for all root-to-tip paths in the tree.</p>
<p>&#40;Recall that an evolutionary distance is a product of the form evolutionary rate \(\times\) time. Keeping dimensions in the back of your head often helps: distance &#91;substitutions/site&#93; &#61; rate &#91;substitutions/site/year&#93; \(\times\) time &#91;year&#93;&#41;</p>
<blockquote>
<p>Question: based on your knowledge of the tree of life, can you identify where the phylogeny is &#40;very likely&#41; wrong?</p>
</blockquote>
<p>The most eye-catching violation of our understanding of evolution of life on earth involves the placement of Anemonia &#40;sea anemone, a cnidarian animal&#41; as sister to the Fungi, with the resulting clade in turn forming a monophyletic group with plants that is sister to the other animals. Our current understanding is that animals and Fungi are however both within the Opisthokonts, which is a sister lineage &#40;in the present taxon sampling&#41; to the Archaeplastida &#40;containing plants and algae&#41;.</p>
<blockquote>
<p>Question: Clustering methods like UPGMA and WPGMA produce rooted trees. Did the clustering algorithm identify the correct root?</p>
</blockquote>
<p>No, it did not. The correct root split for this set of species would be the split of Archaeplastida and Opisthokonta &#40;so algae&#43;plants vs. fungi&#43;animals&#41;.</p>
<blockquote>
<p>Question: How do you interpret the branch lengths, what is the associated length &#39;unit&#39;?</p>
</blockquote>
<p>The branch lengths are in distance units which correpond the expected number of substitutions per site. Distances reflect a statistical estimate of the ‘amount’ of molecular evolution &#40;see also above&#41;.  In the trees constructed using clustering methods, the branch lengths are derived from the pairwise distance estimates using some rule &#40;which differs for instnce for UPGMA and WPGMA&#41; to combine the \(n \times m\) different pairwise estimates between two clades of \(n\) and \(m\) taxa respectively into two branch lengths. In both UPGMA and WPGMA the rule used to combine these pairwise estimates amounts to making an assumption of constant rates of evolution. When interpreting the branch lengths we should take all this into account. Using different clustering methods will result in different internal branch lengths &#40;but not the branch lengths leading to tip nodes&#41;, while being derived from the same pairwise distance matrix.</p>
<h2 id="neighbor-joining"><a href="#neighbor-joining" class="header-anchor">Neighbor-joining</a></h2>
<blockquote>
<p>Question: How can you see at one glance that this is not an ultrametric tree? What does this mean in terms of assumptions on the substitution rate?</p>
</blockquote>
<p>A rooted non-ultrametric tree would be easy to spot because the distances from root to tips would not be equal. However we have an unrooted tree &#40;see fig. 1&#41; in this case. It is not possible to root this tree such that the root-to-tip distances are identical across the tree however, so it cannot be ultrametric. An easy way to see this is that two leaves that have a direct common ancestor have different branch lengths &#40;e.g. Zea and Oryza or Homo and Xenopus&#41;, which can never be the case in an utrametric tree &#40;see the trees for the clustering methods if you are not convinced&#41;.</p>
<blockquote>
<p>Question: Neighbor-joining infers an unrooted tree, can you see how the fact that the tree is unrooted is represented in FigTree? Where should you root this tree &#40;based on your knowledge of the tree of life&#41;? Select the branch where you think the root should be &#40;click on it&#41; and hit the reroot button in figtree to root the tree.</p>
</blockquote>
<p>An unrooted tree is usually represented by a polytomous root &#40;i.e. a trifurcating root node, see fig. 2&#41;. It is important to not interpret this as a true polytomy in the tree &#40;i.e. where an ancestral lineage would have split into tree distinct lineages at the same time&#41;.</p>
<p>The current consensus view is that the Archaeplastida and Ophistokonta both form monophyletic groups, so we should put the root in the branch connecting the last common ancestor of the Ophistokonta &#40;e.g. last common ancestor of <em>Homo</em> and <em>Saccharomyces</em>&#41; and the last common ancestor of Archaeplastida &#40;e.g. last common ancestor of <em>Palmaria</em> and <em>Zea</em>&#41;. See fig. 3. Note that when you root the tree in FigTree, FigTree assumes you want to put the root in the middle of the branch &#40;i.e. the total distance between the two clades stemming from the root is divided into two equal length daughter branches&#41;.</p>
<blockquote>
<p>Question: Does the NJ tree make more sense than the WPGMA tree? What do you think is causing this?</p>
</blockquote>
<p>Clearly it does. Since the pairwise distances are the same for the NJ tree and the WPGMA tree, the difference is due to the tree construction method not the substitution model. Presumably the assumption of equal amounts of evolution &#40;constant rates of evolution&#41; across the phylogeny is false. Some branches in the NJ tree are much further from the root &#40;or some other common ancestral node&#41; than others, indicating unequal rates of evolution across the tree. This is of course not surprising, given that this tree describes the relationships between vastly different organisms, with very different biologies that may affect key parameters of the molecular evolutionary process.</p>
<h2 id="gamma_distances"><a href="#gamma_distances" class="header-anchor">Gamma distances</a></h2>
<blockquote>
<p>Question: interpret the Gamma distribution as a model for rate heterogeneity using the plot above. To what assumption on the rates across sites does a large value of \(\alpha\) correspond? To what assumption does a small value of \(\alpha\) correspond?</p>
</blockquote>
<p>The plot shows the probability density function for the distribution of relative substitution rates across sites. As a model of rate heterogeneity across sites, we can interpret the plot for instance as follows: the probability that a particular site has a relative substitution rate &#40;relative to the other sites in the alignment that is&#41; in the interval \([0.2, 0.8]\) &#40;for instance&#41; is equal to the area under the curve between x-coordinates 0.2 and 0.8 &#40;i.e. the definite integral \(\int_{0.2}^{0.8} f(x)dx\) of the gamma density function \(f(x)\)&#41;.</p>
<p>A large value of \(\alpha\) corresponds to a distribution which is quite narrow around one, so there is not much variability in rates across sites. A small value of \(\alpha\) indicates that there is a strongly asymmetric distribution of rates across sites, with many sites having low relative rates, and many sites having rather high relative rates.</p>
<blockquote>
<p>Question: What &#40;if anything&#41; is changing? What is the default \(\alpha\) used in FastME? Experiment with the \(\alpha\) parameter of the Gamma distribution by using for instance -g0.5 in the FastME command. What happens?</p>
</blockquote>
<p>The branch lengths change considerably, both relative to each other as well as in overall length. Assuming highly asymmetric distributions of the substitution rate across sites &#40;e.g. \(\alpha\) &#61; 0.2, fig. 4&#41; leads to higher distance estimates overall and more outspoken differences between different branches. This is due to the combination of both allowing different substitution rates across branches and across sites, leading to multiplicative effects &#40;in contrast with the clustering methods, where we saw not much change in the relative branch lengths when using \(\Gamma\) heterogeneity&#41;. Using more symmetric distributions &#40;e.g. \(\alpha\) &#61; 10 fig. 5&#41; leads to a result that is indistinguishable from the result without \(\Gamma\) distances. This is because the different rates across sites are all very close to the mean rate, so the model is well-approximated by a model with a single substitution rate for all sites &#40;i.e. the normal JC model&#41;.</p>
<h1 id="maximum_likelihood_phylogeny_inference"><a href="#maximum_likelihood_phylogeny_inference" class="header-anchor">Maximum likelihood phylogeny inference</a></h1>
<h2 id="analysis_under_the_jukes_cantor_model"><a href="#analysis_under_the_jukes_cantor_model" class="header-anchor">Analysis under the Jukes &amp; Cantor model</a></h2>
<blockquote>
<p>Examine the stuff IQ-TREE prints to the screen, can you figure out &#40;roughly&#41; the different steps IQ-TREE uses to find the ML tree?</p>
</blockquote>
<p>Here I’ll highlight some parts of the output. A first informative bit is</p>
<pre><code class="language-md">Alignment has 45 sequences with 1626 columns, 1110 distinct patterns
934 parsimony-informative, 232 singleton sites, 460 constant sites</code></pre>
<p>This shows some features of the data. We have 1626 columns, i.e. data points &#40;recall that we assume independent evolution across sites, so each column of an alignment is a data point from a statistical point of view&#41;. There are 1110 unique columns &#40;or ‘site patterns’ as they tend to be called in phylogenetics&#41;. Note that the likelihood will be the same for identical columns in the alignment, so the program will not compute the log-likelihood for 1626 sites separately, but run the basic likelihood algorithm for the 1110 unique patterns everytime the likelihood &#40;i.e. \(P(\text{data}|\text{tree,parameters})\)&#41; is calculated, saving time. 934 sites are ‘parsimony informative’ which means that for those sites, some topologies would have a higher parsimony score than others.</p>
<p>Next IQ-TREE performs a compositional heterogeneity test. This is a crude statistical test to assess whether nucleotide composition is similar or not across sequences.  For a ‘tree of life’ scale data set like the one here, you’ll see the sequences have quite different base compositions. More info can be found <a href="http://www.iqtree.org/doc/Frequently-Asked-Questions#what-is-the-purpose-of-composition-test">here</a>.</p>
<p>Then the IQ-TREE algorithm starts.</p>
<ul>
<li><p>First a maximum parsimony tree is inferred to use as an initial tree for the ML search.</p>
</li>
<li><p>Next, initial model parameters are estimated by doing ML estimation on the initial parsimony tree.</p>
</li>
<li><p>Then pairwise distances are estimated using the model with the initial parameter estimates.  </p>
</li>
<li><p>Then a Neighbor-joining tree is inferred based on the distance matrix &#40;using the BIONJ algorithm&#41;, and the loglikelihood is computed for the resulting &#40;BIO&#41;NJ tree. This tree is our next ‘initial’ tree.  </p>
</li>
</ul>
<p>IQ-TREE will then generate a candidate tree set based on the initial tree, compute 1the likelihoods, and then proceed with the best scoring trees.  For these trees it will then try to perform random topological rearrangements etc. to search for the best tree &#40;the one with the highest likelihood&#41;.  After that a smaller candidate set is obtained, and IQ-TREE will do a more thorough optimization of the model parameters, branch lengths and topology to arrive at a final tree estimate. For the final tree, it will then do a final maximum likelihood estimation step of the model parameters.</p>
<blockquote>
<p>Open the <code>.iqtree</code> file. What is the log-likelihood associated with the ML tree?</p>
</blockquote>
<p>Check the line with: </p>
<pre><code class="language-md">BEST SCORE FOUND : -31599.974</code></pre>
<blockquote>
<p>Do you think it is a problem that the likelihood is such an inconceivably small number?</p>
</blockquote>
<p>No, this is perfectly expected. Recall that the likelihood L is defined as the probability of the data given the tree topology, model parameters and branch lengths, i.e. </p>
\[\ell(X|\Psi,b,\theta) = P(X|\Psi,b,\theta)\]
<p>where \(X\) denotes the data, \(\Psi\) the topology, \(b\) the branch lengths &#40;i.e. \(\Psi\) and \(b\) together define the <em>phylogenetic tree</em>&#41; and \(\theta\) denotes the parameters of the substitution model.  The likelihood of \(e^{−31600}\) is thus the probability that if you randomly simulate the evolution of a 1626 site long-sequence from the Jukes-Cantor with the inferred ML parameters along the inferred ML tree, you will obtain the exact data set you have used as input. Of course this probability really is ridiculously small. &#40;Back of the envelope: note that the log-likelihood per site is about -19, which you can compare to the log-likelihood of observing a random 45 character string &#40;one nucleotide for each species&#41; for a four-letter alphabet under a uniform distribution, which is \(\log(1/4^{45}) \approx −62\)&#41;.</p>
<blockquote>
<p>How many parameters did we have to estimate?</p>
</blockquote>
<p>Check the line:</p>
<pre><code class="language-md">Number of free parameters &#40;#branches &#43; #model parameters&#41;: 87</code></pre>
<h2 id="gamma_distributed_rates"><a href="#gamma_distributed_rates" class="header-anchor">Gamma distributed rates</a></h2>
<blockquote>
<ul>
<li><p>What is the MLE for the \(\alpha\) parameter of the Gamma distribution?</p>
</li>
<li><p>How many rate classes does IQ-TREE use by default? </p>
</li>
<li><p>What are the relative rates for each rate class for the estimated Gamma distribution? </p>
</li>
<li><p>Is the implied Gamma distribution more asymmetric or less asymmetric then an Exponential distribution?</p>
</li>
</ul>
</blockquote>
<pre><code class="language-md">Model of rate heterogeneity: Gamma with 4 categories
Gamma shape alpha: 0.46

 Category  Relative_rate  Proportion
  1         0.02623        0.25
  2         0.225          0.25
  3         0.7888         0.25
  4         2.96           0.25
Relative rates are computed as MEAN of the portion of the Gamma distribution falling in the category.</code></pre>
<p>The ML estimate is \(\alpha = 0.46\). Note that we approximate the Gamma distribution of rates across sites by fitting a discretized Gamma model with 4 classes. The relative rates associated with the different categories are shown in the .iqtree output. &#40;see the plots and explanation in the exercises for more explanation on the discretization&#41;. Clearly, IQ-TREE uses four rate classes by  default. The relative rates are listed in the table. The distribution is more asymmetric than an Exponential distribution &#40;\(\alpha < 1\), see plots on the exercises page for distance methods&#41;.</p>
<h2 id="other_common_substitution_models"><a href="#other_common_substitution_models" class="header-anchor">Other common substitution models</a></h2>
<blockquote>
<p>What is the difference between the K2P and JC model? </p>
</blockquote>
<p>The K2P model allows for unequal substitution rates for transitions &#40;pyrimidine to pyrimidine, purine to purine&#41; vs.  tranversions &#40;pyrimidine to purine and vice versa&#41;. </p>
<blockquote>
<p>How many more parameters does the K2P model have compared to the JC model? </p>
</blockquote>
<p>The K2P has one additional parameter, the ratio of the substitution rate for transitions to the substitution rate for transversions. This is confirmed by the IQ-TREE output: <code>Number of free parameters &#40;#branches &#43; #model parameters&#41;: 88</code></p>
<blockquote>
<p>What are the ML estimate&#40;s&#41; of the&#40;se&#41; parameter&#40;s&#41;?</p>
</blockquote>
<p>We find in the <code>.iqtree</code> file:</p>
<pre><code class="language-md">Rate parameter R:

  A-C: 1.0000
  A-G: 2.3152
  A-T: 1.0000
  C-G: 1.0000
  C-T: 2.3152
  G-T: 1.0000</code></pre>
<p>IQ-TREE sets the rate of transversions to 1 and estimates the ratio of transition rate to the transversion rate &#40;this parameter is often denoted \(\kappa\)&#41;. Here we estimated \(\hat{\kappa} = 2.3152\), so given that a substitution occurs, it is 2.3152 as likely that it will be a transition compared to a transversion.</p>
<blockquote>
<p>Extra: Consider the F81, HKY and GTR models. Look again at the .iqtree output files. How do these models relate to the JC and K2P model? How do they relate to each other? </p>
</blockquote>
<p>Briefly: the F81 model allows for unequal equilibrium frequencies &#40;K2P and JC assume the equilibrium frequencies – i.e. the base frequencies when the evolutionary process assumed by the model is run for a long time – to be all 0.25&#41; but assumes equal substitution rates for all substitution types. The HKY model allows for both unequal base frequencies and a different rate for transitions compared to transversions, so it can be seen as a combination of F81 and K2P &#40;if you set the equilibrium frequencies in HKY to 0.25 you get K2P, if you set \(\kappa\) to one, you get F81&#41;. The GTR model allows for unequal equilibrium base frequences and allows all types of substitions to occur at different rates &#40;but assumes symmetry, i.e. the rate of <code>A -&gt; T</code> is the same as <code>T -&gt; A</code>&#41;. It has all other substitution models we discussed as a special case.</p>
<h2 id="the_bootstrap"><a href="#the_bootstrap" class="header-anchor">The bootstrap</a></h2>
<blockquote>
<p>What does the <code>-wbtl</code> option do? </p>
</blockquote>
<p>This option writes the inferred trees for the bootstrap replicates to a file with extension <code>.ufboot</code>.</p>
<blockquote>
<p>Interpret how a bootstrap support value is obtained in terms of the <code>.ufboot</code> file </p>
</blockquote>
<p>In the <code>.ufboot</code> file each line is a tree topology obtained for a bootstrapped alignment &#40;see the course on what that means&#41;. A bootstrap support value for a particular clade will be the proportion of bootstrap trees &#40;which are written in that file&#41; which contain that clade. </p>
<p>Note that I say clade and not subtree.  The support value should be interpreted as the support for a bipartition &#40;or <em>split</em>&#41;, not a subtree. To make this clear consider the tree for the 18SrRNA data set with 20 species:</p>
<p><img src="/assets/phylocourse/mliqtree/20species-boot.svg" alt="" /></p>
<p>Take for instance the bootstrap support value of 85 for the clade of green plants. What this indicates is that in the 1000 bootstrap replicates we found that 85&#37; of the <em>unrooted</em> trees contain a clade which consists of the green plants, or in other words, 85&#37; of the trees contain a bipartition &#40;i.e. we can cut the unrooted tree in two trees&#41; such that one of the resulting subtrees consists of the green plants and the other of all the rest. Importantly, this does not mean that 85&#37; of the bootstrap replicates contain the exact <em>subtree</em> we see in this figure, the branching patterns within the subtrees defined by the bipartition may differ.</p>
<div class="page-foot">
  <div class="copyright">
    &copy; Arthur Zwaenepoel. Last modified: May 13, 2021. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>
</div><!-- CONTENT ENDS HERE -->
    
        <script src="/libs/katex/katex.min.js"></script>
<script src="/libs/katex/auto-render.min.js"></script>
<script>renderMathInElement(document.body)</script>

    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
  </body>
</html>
