<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <!-- Begin Jekyll SEO tag v2.4.0 -->
<title>Neutral evolution | bb</title>
<meta name="generator" content="Jekyll v3.7.3" />
<meta property="og:title" content="Neutral evolution" />
<meta name="author" content="Arthur Zwaenepoel" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="So I bought Lynch &amp; Walsh ginormous book ‘Evolution &amp; Selection of Quantitative Traits’, and as I would like to study it thoroughly, I’ll keep something of a study diary on it. This book, with it’s bible-like impression, a book of revelation so it seems, a work of titans deserves a profound study, despite it’s overwhelming magnitude. Indeed, the question is probably rather whether I deserve the book or have the brains for it. Anyway, I’m not discussing the first chapter, although it is very interesting, and here I will be delving in the second chapter, which gives an impression of being quite basic, but nevertheless touches on some advanced matters (for me at least). The chapter is on neutral evolution in one and two-locus systems, and is evidently of great importance for evolutionary genetics in general. It is concerned with genetic drift, mutation and migration." />
<meta property="og:description" content="So I bought Lynch &amp; Walsh ginormous book ‘Evolution &amp; Selection of Quantitative Traits’, and as I would like to study it thoroughly, I’ll keep something of a study diary on it. This book, with it’s bible-like impression, a book of revelation so it seems, a work of titans deserves a profound study, despite it’s overwhelming magnitude. Indeed, the question is probably rather whether I deserve the book or have the brains for it. Anyway, I’m not discussing the first chapter, although it is very interesting, and here I will be delving in the second chapter, which gives an impression of being quite basic, but nevertheless touches on some advanced matters (for me at least). The chapter is on neutral evolution in one and two-locus systems, and is evidently of great importance for evolutionary genetics in general. It is concerned with genetic drift, mutation and migration." />
<link rel="canonical" href="http://localhost:4000/2018/08/10/neutral_evolution.html" />
<meta property="og:url" content="http://localhost:4000/2018/08/10/neutral_evolution.html" />
<meta property="og:site_name" content="bb" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2018-08-10T22:23:32+02:00" />
<script type="application/ld+json">
{"dateModified":"2018-08-10T22:23:32+02:00","datePublished":"2018-08-10T22:23:32+02:00","@type":"BlogPosting","headline":"Neutral evolution","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/2018/08/10/neutral_evolution.html"},"url":"http://localhost:4000/2018/08/10/neutral_evolution.html","author":{"@type":"Person","name":"Arthur Zwaenepoel"},"description":"So I bought Lynch &amp; Walsh ginormous book ‘Evolution &amp; Selection of Quantitative Traits’, and as I would like to study it thoroughly, I’ll keep something of a study diary on it. This book, with it’s bible-like impression, a book of revelation so it seems, a work of titans deserves a profound study, despite it’s overwhelming magnitude. Indeed, the question is probably rather whether I deserve the book or have the brains for it. Anyway, I’m not discussing the first chapter, although it is very interesting, and here I will be delving in the second chapter, which gives an impression of being quite basic, but nevertheless touches on some advanced matters (for me at least). The chapter is on neutral evolution in one and two-locus systems, and is evidently of great importance for evolutionary genetics in general. It is concerned with genetic drift, mutation and migration.","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

  <link rel="stylesheet" href="/assets/main.css">
  <link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="bb" />
</head>


  <body>

    <header class="site-header" role="banner">

  <div class="wrapper">
    
    
    <a class="site-title" rel="author" href="/">bb</a>

    <!--
    
      <nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
              <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
              <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger">
          
            
            
          
            
            
            <a class="page-link" href="/pages/handout.html">Evolutionary analysis - phylogenetic tree inference using maximum likelihood and Bayesian inference</a>
            
          
            
            
          
            
            
          
            
            
          
            
            
            <a class="page-link" href="/mubb.html">mubb</a>
            
          
            
            
          
            
            
          
            
            
          
            
            
          
        </div>
      </nav>
    -->
    
  </div>
</header>


    <main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="w3-col l6 m12 s12">

<article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header" style="text-align:left;">
    <h1 class="post-title p-name" itemprop="name headline">Neutral evolution</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2018-08-10T22:23:32+02:00" itemprop="datePublished">
        
        Aug 10, 2018
      </time>
      
        • <span itemprop="author" itemscope itemtype="http://schema.org/Person"><span class="p-author h-card" itemprop="name">Arthur Zwaenepoel</span></span>
      </p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <p>So I bought Lynch &amp; Walsh ginormous book ‘Evolution &amp; Selection of Quantitative
Traits’, and as I would like to study it <em>thoroughly</em>, I’ll keep something of a
study diary on it.  This book, with it’s bible-like impression, a book of
revelation so it seems, a work of titans deserves a profound study, despite
it’s overwhelming magnitude. Indeed, the question is probably rather whether I
deserve the book or have the brains for it. Anyway, I’m not discussing the
first chapter, although it is very interesting, and here I will be delving in
the second chapter, which gives an impression of being quite basic, but
nevertheless touches on some advanced matters (for me at least). The chapter is
on neutral evolution in one and two-locus systems, and is evidently of great
importance for evolutionary genetics in general. It is concerned with genetic
drift, mutation and migration.</p>

<h2 id="the-wright-fisher-model">The Wright-Fisher model</h2>

<p>The well-known Wright-Fisher model is obviously a good place to start. It
assumes</p>

<ul>
  <li>a constant population size $N$</li>
  <li>monoecious individuals</li>
  <li>random mating</li>
  <li>discrete generations</li>
</ul>

<p>and obviously for now, no selection, mutation or migration. Consider one
allele/gene in this population, present in $i$ copies, the probability of
having $j$ copies in the next generation is given by the binomial distribution
$Binom(n, p)$ with $n = 2N$ (that is the number of genes to sample to get the
next generation) and $p = i/2N$ (the frequency of the gene we are following).
Specifically, we have</p>

<script type="math/tex; mode=display">P_{ij} = \binom{2N}{j} \Big(\frac{i}{2N}\Big)^{j}\Big(1 -
\frac{i}{2N}\Big)^{2N-j}</script>

<p>This $[P_{ij}]$ values expressing the probability of transition from copy
number $i$ to $j$ form a transition probability matrix $P$ with dimensions
$(2N+1) \times (2N+1)$. Note that all $P_{0.}$ values except $P_{00}$ are 0 and
likewise for $P_{1.}$. Also note that all rows sum to one. The probability
distribution over the copy number of the allele of interest at time $t$ is
given by</p>

<script type="math/tex; mode=display">x(t) = x(0)P^t</script>

<p>with $x$ a probability vector for the number of copies of the allele under
consideration. As indicated, there are two absorbing states in the model,
since there is no mutation or migration there is no source of novel variation
and an allele will either get lost or fixed eventually. In terms of the probability
vector over copy number values $x$, this translates to saying that the sum
$x_{2N} + x_{0}$ converges to one as $t \rightarrow \infty$.</p>

<p>The variance in alele frequencies attained conditional on the allel frequency
in the parental population can serve as a measure for the dispersion in allele
frequency induced by random sampling in the Wright-Fisher model. After one
generation of random mating the expected number of allel copies in the
population is given by the Hardy-Weinberg proportions, <em>i.e.</em></p>

<script type="math/tex; mode=display">\mathbb{E}(N) = p^2 \times 2 + 2p(1-p) \times 1 + (1-p)^2\times 0 = 2p</script>

<p>and likewise</p>

<script type="math/tex; mode=display">\mathbb{E}(N^2) = p^2 \times 4 + 2p(1-p)\times 1 = 2p(p+1)</script>

<p>giving</p>

<script type="math/tex; mode=display">Var(n) = 2p(p+1) - 4p^2 = -2p^2 +2p = 2p(1-p)</script>

<p>where $n$ is for copy number. To be clear, this is the variance in the number
of copies of the allele of interest in an individual. To get the variance for
the allele frequency in the population, we need the variance in the total
number of copies in the population divde by the number of alleles in the
population, that is</p>

<script type="math/tex; mode=display">NVar\Big(\frac{n}{2N}\Big) = \frac{2Np(1-p)}{(2N)^2} = \frac{p(1-p)}{2N}</script>

<p>Which is the sampling variance of a binomial random variable $Binom(2N, p)$.</p>

<p>So why is this interesting? It shows that the expected fraction of
heterozygotes is a measure forthe variance in allele frequency, and
that we cn track heterozygosity to get  grip on the effects of drift
on the variation in the population.</p>

<h2 id="loss-of-heterozygosity-in-a-wright-fisher-population">Loss of heterozygosity in a Wright-Fisher population</h2>

<p>Now that we know that heterozygosity can serve as a measure for the effect of
drift in a Wright-Fisher population, it might be interesting to study this in
more detail. Considering an initially fully unrelated population, the probability
that two genes are identical by descent in the next generation is given by</p>

<script type="math/tex; mode=display">f(1) = \frac{1}{2N} + \Big(1 - \frac{1}{2N}\Big)</script>

<p>or in an arbitrary generation</p>

<script type="math/tex; mode=display">f(t) = \frac{1}{2N} + \Big(1 - \frac{1}{2N}\Big)f(t-1)</script>

<p>This is of course the expected inbreeding coefficient in generation $t$ under
random-mating in a Wright-Fisher population of size N. This is closely related to the heterozygosity in the population. Indeed, an individual can only be a heterozygous for a locus if the alleles are not IBD, which occurs with probability $1-f$. In a random mating population, the fraction of heterozygotes $2p(1-p)$ is reduced proportional to the amount of inbreeding in the population. This fractional reduction $1-f$ we term the heterozygosity, for which we can find a nice recursion:</p>

<script type="math/tex; mode=display">% <![CDATA[
\begin{eqnarray}
H(t) &=& 1 - \frac{1}{2N} - \Big(1-\frac{1}{2N}\Big)f(t-1) \\
     &=& \Big(1 - \frac{1}{2N}\Big)H(t-1) \\  
     &=& \Big(1 - \frac{1}{2N}\Big)^tH(0)
\end{eqnarray} %]]></script>

<p>giving an expression for the <em>expected heterozygosity</em> at time $t$ relative to some base population with heterozygosity $H(0)$.</p>

<script type="math/tex; mode=display">H(t) \approx e^{\frac{-t}{2N}}H(0)</script>

<p>This expression can be used to calculate expected times for particular reductions etc.</p>

  </div>

  

  <div class="post-categories">
    
    
  </div>
  <a class="u-url" href="/2018/08/10/neutral_evolution.html" hidden></a>
</article>

</div>

<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript" ></script>
<script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: { autoNumber: "AMS" } } }); </script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>

      </div>
    </main>


  </body>

</html>
