<!doctype html> <html lang=en > <meta charset=UTF-8 > <meta name=viewport content="width=device-width, initial-scale=1"> <link rel=stylesheet href="/libs/katex/katex.min.css"> <link rel=stylesheet href="/libs/highlight/github.min.css"> <link rel=stylesheet href="/css/franklin.css"> <link rel=stylesheet href="/css/basic.css"> <title>SARS-CoV-2 phylogenetics</title> <header> <div class=blog-name ><a href="/">bb</a></div> <nav> <ul> </ul> </nav> </header> <div class=franklin-content > <div class=franklin-toc ><ol><li><a href="#obtaining_data">Obtaining data</a><ol><li><a href="#viral_genomes">Viral genomes</a><li><a href="#gene_families">Gene families</a></ol><li><a href="#multiple_sequence_alignment">Multiple sequence alignment</a><li><a href="#phylogenetic_inference_using_maximum_likelihood">Phylogenetic inference using Maximum likelihood</a></ol></div> <p>In this exercise, we will have a look at the phylogenetic position of the human coronavirus SARS-CoV-2 that is currently causing the outbreak of COVID-19. The goal of this exercise is to go through a complete workflow for phylogenetic analysis.</p> <p><strong>Disclaimer:</strong> I am in no way experienced with viral phylogenomics. I know nothing about the available data, little about viral evolutionary relationships, and barely anything about viral genomes. Also the data used below is in no way optimal for addressing questions about SARS-CoV-2 evolution, for which the <a href="https://www.gisaid.org/">GISAID</a> database is probably more appropriate &#40;but requires registration and prohibits secondary sharing of dt with unregistered individuals&#41;.</p> <h2 id=obtaining_data ><a href="#obtaining_data">Obtaining data</a></h2> <h3 id=viral_genomes ><a href="#viral_genomes">Viral genomes</a></h3> <p>As armchair biologists, we tend not to go look for our data out there in the wild &#40;especially during a pandemic&#41;, but download it from some database. Here I describe how I got the data set we will be analyzing later. This is not really part of the practical, but rather then providing you a data set rightaway, it can be informative to see the typical annoying bits of bioinformatics that necessarily come first.</p> <p>To get the data, I first did a search in the NCBI &#39;Nucleotide&#39; database using the following search term:</p> <pre><code class=language-julia >("Alphacoronavirus"[Organism] OR "Betacoronavirus"[Organism] OR
 "Gammacoronavirus"[Organism]) AND (viruses[filter] AND refseq[filter])</code></pre> <p>I restrict the sequence data to the RefSeq database, which is a collection of non-redundant and well-annotated sequences. As I am no expert in viral phylogenomics, I considered this the best idea, since it would not be feasible for me to assess data quality and redundancy efficiently. This provided me with 44 corona virus genome sequences.</p> <p>I sent the results to a file, using the &#39;Complete record&#39; option and the GenBank format. Note that the reference sequence for nCov-2019 is the Wuhan isolate apparently, with the following <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512.2?report&#61;genbank&amp;to&#61;29903">genbank entry</a>.</p> <p>Now we could work with the entire genomes, aligning them and trying to do tree inference for the full viral genome alignments. Alternatively, we could analyze the different viral proteins separetely. Besides being computationally less involved, this has also the advantage of being somewhat clearer to work with. I wrote the following little script using Biopython to obtain the coding DNA sequences &#40;CDS&#41; and associated amino-acid &#40;AA&#41; sequences for each viral genome:</p> <pre><code class=language-py >from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
import os
import sys

if len(sys.argv) != 3:
	print("usage: {} [genbank file] [prefix]".format(sys.argv[0]))
    exit()

path_nt = sys.argv[2] + "-nt"
path_aa = sys.argv[2] + "-aa"
os.mkdir(path_nt)
os.mkdir(path_aa)
for rec in SeqIO.parse(sys.argv[1], "genbank"):
    if rec.features:
        gid = rec.annotations["organism"].replace(" ", "_").replace("/", "_")
        recs = [SeqRecord(
            seq=ft.location.extract(rec).seq,
            id=ft.qualifiers['protein_id'][0],
            description="") for ft in rec.features if ft.type == "CDS"]
        SeqIO.write(recs, os.path.join(path_nt, gid + ".fasta"), "fasta")
        trrec = [r.translate(id=r.id, description=r.description) for r in recs if len(r.seq)%3==0]
        SeqIO.write(trrec, os.path.join(path_aa, gid + ".fasta"), "fasta")</code></pre> <p>The script above can be run as follows:</p> <pre><code class=language-julia >python3 getdata.py sequence.gb genomes</code></pre>
<p>Which gives us two directories, one with all CDS sequences for each genome and one with all associated amino-acid sequences.</p>
<pre><code class=language-julia >tree genomes-aa

genomes-aa
├── Bat_coronavirus_1A.fasta
├── Bat_coronavirus_CDPHE15_USA_2006.fasta
├── Bat_Hp-betacoronavirus_Zhejiang2013.fasta
├── Beluga_whale_coronavirus_SW1.fasta
├── Betacoronavirus_England_1.fasta
├── Betacoronavirus_Erinaceus_VMC_DEU_2012.fasta
├── Betacoronavirus_HKU24.fasta
├── Bovine_coronavirus.fasta
├── BtMr-AlphaCoV_SAX2011.fasta
├── BtNv-AlphaCoV_SC2013.fasta
├── BtRf-AlphaCoV_HuB2013.fasta
├── BtRf-AlphaCoV_YN2012.fasta
├── Camel_alphacoronavirus.fasta
├── Coronavirus_AcCoV-JC34.fasta
├── Feline_infectious_peritonitis_virus.fasta
├── Ferret_coronavirus.fasta
├── Human_coronavirus_229E.fasta
├── Human_coronavirus_HKU1.fasta
├── Human_coronavirus_NL63.fasta
├── Human_coronavirus_OC43.fasta
├── Infectious_bronchitis_virus.fasta
├── Lucheng_Rn_rat_coronavirus.fasta
├── Middle_East_respiratory_syndrome-related_coronavirus.fasta
├── Miniopterus_bat_coronavirus_HKU8.fasta
├── Mink_coronavirus_strain_WD1127.fasta
├── Murine_hepatitis_virus.fasta
├── Murine_hepatitis_virus_strain_JHM.fasta
├── NL63-related_bat_coronavirus.fasta
├── Pipistrellus_bat_coronavirus_HKU5.fasta
├── Porcine_epidemic_diarrhea_virus.fasta
├── Rabbit_coronavirus_HKU14.fasta
├── Rat_coronavirus_Parker.fasta
├── Rhinolophus_bat_coronavirus_HKU2.fasta
├── Rousettus_bat_coronavirus.fasta
├── Rousettus_bat_coronavirus_HKU10.fasta
├── Rousettus_bat_coronavirus_HKU9.fasta
├── Scotophilus_bat_coronavirus_512.fasta
├── Severe_acute_respiratory_syndrome_coronavirus_2.fasta
├── Severe_acute_respiratory_syndrome-related_coronavirus.fasta
├── Swine_enteric_coronavirus.fasta
├── Transmissible_gastroenteritis_virus.fasta
├── Turkey_coronavirus.fasta
├── Tylonycteris_bat_coronavirus_HKU4.fasta
└── Wencheng_Sm_shrew_coronavirus.fasta</code></pre>
<p>The genomes seem to have between 5 and 14 genes, as can be quickly inspected using <code>grep -c &#39;&gt;&#39; genomes-aa/* | cut -d&#39;:&#39; -f2 | sort | uniq</code>.</p>
<h3 id=gene_families ><a href="#gene_families">Gene families</a></h3>
<p>With these viral genomes nicely in their respective files, we do not have our gene families of interest yet. Since these are RefSeq records, we could probably get gene families &#40;e.g. the S-protein&#41; for each genome directly by looking at the gene names, but this is tricky &#40;what if the names are not consistent for instance?&#41;. I&#39;d rather take a more general approach. We&#39;ll make a blast database for our genomes, and fish out sequences using a bait gene of interest &#40;for instance the S-protein from SARS-CoV-2&#41;.</p>
<p>Well, actually, I&#39;ll use <a href="https://github.com/bbuchfink/diamond">diamond</a> instead of blast, which is way faster. First I make a diamond database:</p>
<pre><code class=language-julia >cat genomes-aa/* > aa.fa
diamond makedb --in aa.fa --db aa</code></pre>
<p>Now we fish for the S-protein &#40;spike glycoprotein&#41;. This is the sequence of the Wuhan isolate reference genome of SARS-Cov-2 &#40;which you can easily parse out the genbank entry linked to above&#41;:</p>
<pre><code class=language-julia >>spike-SARS-CoV-2
MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFS
NVTWFHAIHVSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIV
NNATNVVIKVCEFQFCNDPFLGVYYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLE
GKQGNFKNLREFVFKNIDGYFKIYSKHTPINLVRDLPQGFSALEPLVDLPIGINITRFQT
LLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETK
CTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISN
CVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIAD
YNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPC
NGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVN
FNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITP
GTNTSNQVAVLYQDVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSY
ECDIPIGAGICASYQTQTNSPRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTI
SVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQE
VFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDC
LGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAM
QMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALN
TLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRA
SANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPA
ICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDP
LQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDL
QELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDD
SEPVLKGVKLHYT</code></pre>
<p>Run diamond using the above protein as query:</p>
<pre><code class=language-julia >diamond blastp --db aa.dmnd -q spike-sarscov2.fasta > spike.hits</code></pre>
<p>Now we obtain the relevant sequences using the following script &#40;I saved it as <code>gethits.py</code>&#41;</p>
<pre><code class=language-julia >import sys

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("usage: {} <hits> <fastadir>".format(sys.argv[0]))
        exit()
    from Bio import SeqIO
    import pandas as pd
    import os
    import logging
    df = pd.read_csv(sys.argv[1], sep="\t", header=None)
    seqs = {}
    for f in os.listdir(sys.argv[2]):
        for seq in SeqIO.parse(os.path.join(sys.argv[2], f), "fasta"):
            if seq.id in list(df[1]):
                if seq[-1] == "*": seq = seq[0:-1]
                newid = "{}__{}".format(f.split(".")[0], seq.id)
                if newid in seqs: logging.warning("duplicate!")
                seqs[newid] = seq        
    if len(seqs) < len(df.index):
        logging.warning("Found {}/{} sequences".format(
			len(seqs), len(df.index)))
    for (k,v) in seqs.items():
        print(">{}".format(k))
        print(v.seq)</code></pre>
<p>Use it as follows:</p>
<pre><code class=language-julia >python3 gethits.py ./spike.hits ./genomes-aa > spike.aa.fasta</code></pre>
<p>and to get the CDS sequences:</p>
<pre><code class=language-julia >python3 gethits.py ./spike.hits ./genomes-nt > spike.nt.fasta</code></pre>
<p>Clearly, you can use this approach to get other genes from the viral genome database we assembled. Now we can do phylogenetics using the <code>spike.aa.fasta</code> and/or <code>spike.nt.fasta</code> files.</p>
<h2 id=multiple_sequence_alignment ><a href="#multiple_sequence_alignment">Multiple sequence alignment</a></h2>
<p>In order to do phylogenetic inference, we need to have a hypothesis of homology, that is to say, we need to know which characters in a bunch of sequences trace back to a common ancestral character. Sadly, this is not given, and we have to infer this from the data. This is of course the problem of multiple sequence alignment &#40;MSA&#41;.</p>
<p>In practice, inferring a MSA couldn&#39;t be easier:</p>
<pre><code class=language-julia >mafft spike.aa.fasta > spike.aa.msa</code></pre>
<p>And we can have a look at this alignment using for instance the <a href="https://github.com/arvestad/alv"><code>alv</code></a> program. Here&#39;s a little piece of the output of <code>alv spike.aa.msa</code>:</p>
<pre><code class=language-julia >Human_coronavirus_NL63__YP_003767.1                                  --IISVVFVVLLSLLVFCCLSTGC-CGCCNCLTSSMRGCCDC-----
Coronavirus_AcCoV-JC34__YP_009380521.1                               --AIFLVLVLFSFMLLWCCCATGC-CGCCGMLGSACNGCCTK-----
Murine_hepatitis_virus__NP_045300.1                                  --LIGLAGVAVCVLLFFICCCTGCGSCCFK----KCGNCCDE-----
Rat_coronavirus_Parker__YP_003029848.1                               --LIGLAGVAVCVLLFFICCCTGCGSCCFK----KCGNCCDE-----
Infectious_bronchitis_virus__NP_040831.1                             --AIAFATIIFILILGWVFFMTGC-CGC----------CCGCFGIMP
Pipistrellus_bat_coronavirus_HKU5__YP_001039962.1                    GFIAGLVALALCVFFIL--CCTGCGTSCLGKL--KCNRCCDS-----
Severe_acute_respiratory_syndrome_coronavirus_2__YP_009724390.1      GFIAGLIAIVMVTIMLC--CMTSC-CSCLKGC-CSCGSCC-K-----
Rousettus_bat_coronavirus__YP_009273005.1                            GMIAGLVGLVLAVVLLL--CMTNC-CSCARGV-CSCKSC--A-----
Betacoronavirus_Erinaceus_VMC_DEU_2012__YP_009513010.1               GFIAGLVALALCLFFIL--CCTGCGTSCLGKI--NCTRCCDK-----
Severe_acute_respiratory_syndrome-related_coronavirus__NP_828851.1   GFIAGLIAIVMVTILLC--CMTSC-CSCLKGA-CSCGSCC-K-----
Human_coronavirus_HKU1__YP_173238.1                                  --LISFSFIIFLVLLFFICCCTGCGSACFS----KCHNCCDE-----
Lucheng_Rn_rat_coronavirus__YP_009336484.1                           --AIFLVLVLFSFLMLWCCCATGC-CGCCGLCGAACNGCCTK-----
Betacoronavirus_England_1__YP_007188579.1                            GFIAGLVALALCVFFIL--CCTGCGTNCMGKL--KCNRCCDR-----
Bovine_coronavirus__NP_150077.1                                      --LIGFAGVAMLVLLFFICCCTGCGTSCFK----KCGGCCDD-----
Bat_Hp-betacoronavirus_Zhejiang2013__YP_009072440.1                  GFIAGLIAIVMATIMLC--CMTSC-CSCFKGL-CACKRCCDS-----
Rabbit_coronavirus_HKU14__YP_005454245.1                             --LIGLAGVAVLVLLFFICCCTGCGTSCFK----KCGGCCDD-----
Middle_East_respiratory_syndrome-related_coronavirus__YP_009047204.1 GFIAGLVALALCVFFIL--CCTGCGTNCMGKL--KCNRCCDR-----
Human_coronavirus_OC43__YP_009555241.1                               --LICLAGVAMLVLLFFICCCTGCGTSCFK----KCGGCCDD-----
Turkey_coronavirus__YP_001941166.1                                   --AIGFACLIFILILGWVFFMTGC-CGC----------CCGCFGIIP
Rousettus_bat_coronavirus_HKU9__YP_001039971.1                       AMIAGIVGLVLAVIMLM--CMTNC-CSCFKGM-CDCRRCCGS-----
Tylonycteris_bat_coronavirus_HKU4__YP_001039953.1                    GFIAGLVALLLCVFFLL--CCTGCGTSCLGKM--KCKNCCDS-----
Human_coronavirus_229E__NP_073551.1                                  --CISVVLIFVVSMLLLCCCSTGC-CGFFSCFASSIRGCCE------
Murine_hepatitis_virus_strain_JHM__YP_209233.1                       --LIGLAGVAVCVLLFFICCCTGCGSCCFR----KCGSCCDE-----
BtMr-AlphaCoV_SAX2011__YP_009199609.1                                --IIVLVLILFTCLMLFCCCSTGC-CGIFSCMASSCGACCDI-----
Betacoronavirus_HKU24__YP_009113025.1                                --LIGLAGVAVLVLLFFVCCCTGCGSSCFK----KCGSCCDD-----</code></pre>
<p>This however hides the crucial fact that <em>multiple sequence alignment is one of the most important open problems in statistical phylogenetics</em>. Since every inference is associated with uncertainty, ideally we&#39;d like to obtain an idea of the uncertainty in our inferred MSA, and take this uncertainty into account in our phylogenetic analyses. In a Bayesian framework this is possible by jointly inferring the alignment and phylogeny under a model of sequence evolution that includes insertions and deletions &#40;the TKF or Thorne - Kishino - Felsenstein model for instance&#41;, which is what the program <a href="http://bali-phy.org/">BAli-Phy</a> does. However this is computationally very, very expensive.</p>
<p>In the following we will rely on our MAFFT-based MSAs, but it is important to keep  in mind that in all the following statistical phylogenetic analyses, we <em>are making the terrible assumption that the MSA is known without error.</em></p>
<h2 id=phylogenetic_inference_using_maximum_likelihood ><a href="#phylogenetic_inference_using_maximum_likelihood">Phylogenetic inference using Maximum likelihood</a></h2>
<div class=page-foot >
  <div class=copyright >
    &copy; Arthur Zwaenepoel. Last modified: April 12, 2020. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>

</div>

    
        



    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>