<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="stylesheet" href="/libs/katex/katex.min.css">
     
   <link rel="stylesheet" href="/libs/highlight/github.min.css">
   
  <link rel="stylesheet" href="/css/franklin.css">
  <link rel="stylesheet" href="/css/basic.css">
   <title>Distance-based phylogenetic inference</title>  
</head>
<body>
  <header>
	  <!--div class="blog-name"><a href="/">bb</a></div--->
  </header>

<!-- Content appended here -->
<div class="franklin-content"><p><a href="/phylocourse/">back</a> <div class="franklin-toc"><ol><li><a href="#distance-based_phylogenetic_inference">Distance-based phylogenetic inference</a><ol><li><a href="#software_and_data">Software and data</a></li><li><a href="#computing_distance_matrices_using_fastme">Computing distance matrices using FastME</a></li><li><a href="#clustering_methods">Clustering methods</a></li><li><a href="#neighbor-joining">Neighbor-joining</a></li><li><a href="#extra_implementing_neighbor-joining">Extra: implementing Neighbor-Joining</a></li></ol></li></ol></div></p>
<h1 id="distance-based_phylogenetic_inference"><a href="#distance-based_phylogenetic_inference" class="header-anchor">Distance-based phylogenetic inference</a></h1>
<blockquote>
<p><strong>Questions</strong> and <strong>Exercises</strong> will be marked in blocks like this, try to formulate a brief answer to them.</p>
</blockquote>
<blockquote>
<p><strong>Note</strong> that there are some blocks with <code>julia</code> code included below. This is just for illustrating things in a way such that those who are interested in that can actually see how these illustrations &#40;numerical or visual&#41; were actually generated. Don&#39;t let them scare you.</p>
</blockquote>
<p>Distance-based phylogenetic methods proceed by computing a <strong>pairwise distance matrix</strong> for an input alignment based on a <a href="../submod">substitution model</a>, i.e. an assumed <strong>model of evolution</strong> for molecular sequences. Usually these estimated distances are the maximum likelihood estimates &#40;MLE&#41; of the pairwise distance under the assumed substitution model. For instance, consider two sequences, for an observed proportion of different sites \(x\), the maximum likelihood estimate of the distance between the two sequences under the Jukes &amp; Cantor substitution model is</p>
\[\hat{d} = -\frac{3}{4} \log \Big(1 - \frac{4}{3}x\Big)\]
<p>So for instance for the alignment</p>
<pre><code class="language-julia">seqa &#61; &quot;ATCGGGCTAGC&quot;
seqb &#61; &quot;TTCGGCTTACC&quot;;</code></pre>
<p>The proportion of different sites is</p>
<pre><code class="language-julia">differences &#61; sum&#40;&#91;seqa&#91;i&#93; &#33;&#61; seqb&#91;i&#93; for i&#61;1:length&#40;seqa&#41;&#93;&#41;
x &#61; differences/length&#40;seqa&#41;</code></pre><pre><code class="plaintext">0.36363636363636365</code></pre>
<p>The estimated distance is</p>
<pre><code class="language-julia">d &#61; -0.75 * log&#40;1. - 4x/3&#41;</code></pre><pre><code class="plaintext">0.49747066305769816</code></pre>
<p>This distance is correctly interpreted as <strong>an estimate of the expected number of substitutions per site</strong> separating the two sequences. Note that the distance is a product of two other evolutionary variables: the <strong>substitution rate</strong> and the <strong>divergence time</strong>. If we have \(n\) sequences, we can apply the distance formula to obtain an \(n \times n\)  <strong>distance matrix</strong>. Note however that such simple distance formulae are generally not available for more complicated substitution models. For more details, see the <a href="../submod">notes on substitution models</a>.</p>
<blockquote>
<p><strong>Question</strong>. The proportion of different sites in a pair of sequence \(x\) is a <strong>summary statistic</strong> of the data, we reduce the data to a single number \(x\). For the Jukes - Cantor model the likelihood \(P(\text{sequence pair}|\mathrm{model})\) only depends on the summary statistic \(x\). Why will such a summary statistic not work for more complicated substitution models, for instance where transitions and transversions have different rates? If we cannot use a formula based on some summary statistic, how can we estimate a distance?</p>
</blockquote>
<p>After computing the distance matrix, a tree is inferred by using some sort of clustering algorithm or least-squares estimation step. Distance-based phylogenetic inference is thus essentially a two-step process:</p>
<ol>
<li><p>Compute pairwise distances</p>
</li>
<li><p>Infer the tree, assuming the distances</p>
</li>
</ol>
<p>This two step procedure is both the strength and weakness of distance-based methods. By collapsing the sequence data with \(n\) sequences and \(m\) sites in a single \(n \times n\) matrix it dramatically reduces the data, making tree inference computationally very fast. However, by reducing the rich sequence data to a matrix of numbers it throws away a lot of potentially interesting evolutionary information. Another issue is that when using distance-based methods the estimated distances from step 1 <em>are treated as observed data</em> in step 2. However, distances are themselves estimates, associated with some uncertainty &#40;a distance estimate has a variance for instance&#41;, and this uncertainty in the distances is neglected in step 2. Both of these issues are solved in ML and Bayesian phylogenetic inference, however at the price of a strongly increased computational cost.</p>
<p>Because of their speed, distance methods are still quite often used. Some ML tree inference programs &#40;for instance IQ-TREE, see the ML tutorial&#41; start their tree search algorithm from a distance-based tree for instance. Also popular packages for phylogenomic and comparative genomic inference such as <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a> use distance-based phylogenetic methods in their workflow.</p>
<h2 id="software_and_data"><a href="#software_and_data" class="header-anchor">Software and data</a></h2>
<ol>
<li><p>Download the <strong>FastME</strong> software at <a href="http://www.atgc-montpellier.fr/fastme/binaries.php">http://www.atgc-montpellier.fr/fastme</a> &#40;on the bottom of the page, click the download button&#41;. In the downloaded folder you will find a binaries directory, identify the binary &#40;executable&#41; for your operating system and put it in some folder of your convenience.<sup id="fnref:fastmeonline"><a href="#fndef:fastmeonline" class="fnref">[1]</a></sup> For some additional guidance with installing FastME on Windows, have a look <a href="/phylocourse/install-windows">here</a>.</p>
</li>
<li><p>To view trees, I recommend the <a href="https://github.com/rambaut/figtree/releases/tag/v1.4.4">FigTree</a> tool. FigTree requires Java, but that should be available on most machines. Download the executable for your operating system from the link above &#40;<code>.zip</code> file for Windows users, <code>.tgz</code> for *nix users, I guess the <code>.dmg</code> file is something for Mac OSX users&#40;?&#41;&#41;</p>
</li>
<li><p>We will use two small &#39;tree of life&#39; data sets. One <a href="/assets/phylocourse/data/18SrRNA_20.phy">18SrRNA data set with 20 taxa</a> and another <a href="/assets/phylocourse/data/18SrRNA_45.phy">18SrRNA data set with 45 taxa</a>.<sup id="fnref:ancientdata"><a href="#fndef:ancientdata" class="fnref">[2]</a></sup></p>
</li>
</ol>
<h2 id="computing_distance_matrices_using_fastme"><a href="#computing_distance_matrices_using_fastme" class="header-anchor">Computing distance matrices using FastME</a></h2>
<p>While you could easily implement a little program to generate a distance matrix under the Jukes and Cantor model &#40;see the formula&#39;s for the distance in the section on <a href="../submod">substitution models</a>&#41;, this is less straightforward when employing more complicated substitution models. FastME is probably the fastest implementation of distance-based phylogenetics methods using general substitution models available today. It is a software tool that can both be used to compute distance matrices and infer trees using <strong>Neighbor-Joining</strong>, <strong>least-squares</strong>, <strong>minimum evolution</strong> and related distance matrix based methods.</p>
<p>There are two ways to run FastME. There is an interactive mode &#40;inherited from the influential <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP software package</a>&#41; and a command line mode. Personally, I find the command line mode much more convenient. First get a look at the help message</p>
<pre><code class="language-julia">fastme -help</code></pre>
<p>If you have read the course notes from prof. Van de Peer, most of the options listed there &#40;but not all, no worries&#41; should make some sense to you. To infer a distance matrix with the Jukes &amp; Cantor model you can run something like this from the command line<sup id="fnref:commandline"><a href="#fndef:commandline" class="fnref">[3]</a></sup> :</p>
<pre><code class="language-julia">fastme -i 18SrRNA_20.phy -O 18SrRNA_20_JCmatrix.txt -dJC69</code></pre>
<p>Try to run this command and see whether you manage to make it work. The above command should generate a distance matrix. Let&#39;s make a heatmap to visualize it:</p>
<pre><code class="language-julia">using Plots
function readmatrix&#40;file&#41;  # little function to read the FastME distance matrix
    lines &#61; &#91;split&#40;l&#41; for l in readlines&#40;file&#41;&#91;2:end&#93; if l &#33;&#61; &quot;&quot;&#93;
    matrix &#61; hcat&#40;&#91;map&#40;x-&gt;parse&#40;Float64, x&#41;, l&#91;2:end&#93;&#41; for l in lines&#93;...&#41;
    names &#61; &#91;l&#91;1&#93; for l in lines&#93;
    matrix, names, length&#40;names&#41;
end

matrix, taxa, ntaxa &#61; readmatrix&#40;&quot;_assets/phylocourse/distance/18SrRNA_20_JCmatrix.txt&quot;&#41;
heatmap&#40;matrix, yticks&#61;&#40;1:ntaxa, taxa&#41;, xticks&#61;&#40;1:ntaxa, taxa&#41;, xrotation&#61;45, size&#61;&#40;700,650&#41;&#41;</code></pre>
<p><img src="/assets/phylocourse/distance/hm1.svg" alt="" /></p>
<blockquote>
<p><strong>Question</strong>: does this distance matrix make sense to you? Can you spot clades of more closely related species already?</p>
</blockquote>
<h2 id="clustering_methods"><a href="#clustering_methods" class="header-anchor">Clustering methods</a></h2>
<p><strong>Clustering methods</strong> in distance-based phylogenetics are not different from hierarchical clustering methods used in other applications, such as unsupervised machine learning for instance. Both the UPGMA and WPGMA clustering methods are specific cases of <em>average linkage clustering</em> &#40;with different formula&#39;s used to compute the distance between two already existing clusters&#41;. Clustering methods can befound in many packages for scientific computing for most programming languages &#40;here I&#39;ll use the <code>Clustering.jl</code> package for the julia progamming language, note that average linkage clustering as it is usually implemented is identical to what is called the WPGMA method in phylogenetics&#41;.</p>
<pre><code class="language-julia">using Clustering, StatsPlots
hcl &#61; hclust&#40;matrix, linkage&#61;:average&#41;
plot&#40;
    plot&#40;hcl, xticks&#61;false&#41;,
    heatmap&#40;matrix&#91;hcl.order,hcl.order&#93;, colorbar&#61;false,
        yticks&#61;&#40;1:ntaxa, &#91;taxa&#91;i&#93; for i in hcl.order&#93;&#41;,
        xticks&#61;&#40;1:ntaxa, &#91;taxa&#91;i&#93; for i in hcl.order&#93;&#41;,
        xrotation&#61;45&#41;,
    layout&#61;grid&#40;2, 1, heights&#61;&#91;0.2,0.8&#93;&#41;, size&#61;&#40;600,750&#41;&#41;</code></pre>
<p><img src="/assets/phylocourse/distance/wpgma.svg" alt="" /></p>
<p>A pretty figure in my humble opinion.</p>
<blockquote>
<p><strong>Question</strong>: how can you see in one glance that this is an ultrametric tree?</p>
</blockquote>
<blockquote>
<p><strong>Question</strong>: based on your knowledge of the tree of life, can you identify where the phylogeny is &#40;very likely&#41; wrong?</p>
</blockquote>
<blockquote>
<p><strong>Question</strong>: Clustering methods like UPGMA and WPGMA produce rooted trees. Did the clustering algorithm identify the correct root?</p>
</blockquote>
<blockquote>
<p><strong>Question</strong>: How do you interpret the branch lengths, what is the associated length &#39;unit&#39;?</p>
</blockquote>
<p>Now let&#39;s infer a tree using \(\Gamma\)<strong>distributed rates across sites</strong> &#40;i.e. \(\Gamma\) or &#39;Gamma&#39; distances&#41;, still using the Jukes-Cantor substitution model.</p>
<pre><code class="language-julia">fastme -i 18SrRNA_20.phy -O 18SrRNA_20_JCGamma_matrix.txt -dJC69 -g</code></pre>
<pre><code class="language-julia">matrix, taxa, ntaxa &#61; readmatrix&#40;&quot;_assets/phylocourse/distance/18SrRNA_20_JCGamma_matrix.txt&quot;&#41;
hcl &#61; hclust&#40;matrix, linkage&#61;:average&#41;

plot&#40;
    plot&#40;hcl, xticks&#61;false&#41;,
    heatmap&#40;matrix&#91;hcl.order,hcl.order&#93;, colorbar&#61;false,
        yticks&#61;&#40;1:ntaxa, &#91;taxa&#91;i&#93; for i in hcl.order&#93;&#41;,
        xticks&#61;&#40;1:ntaxa, &#91;taxa&#91;i&#93; for i in hcl.order&#93;&#41;,
        xrotation&#61;45&#41;,
    layout&#61;grid&#40;2, 1, heights&#61;&#91;0.2,0.8&#93;&#41;, size&#61;&#40;600,750&#41;&#41;</code></pre>
<p><img src="/assets/phylocourse/distance/wpgma2.svg" alt="" /></p>
<blockquote>
<p><strong>Question</strong>: did the topology change? Did the branch lengths change? Why?</p>
</blockquote>
<h2 id="neighbor-joining"><a href="#neighbor-joining" class="header-anchor">Neighbor-joining</a></h2>
<p>Neighbor-joining &#40;NJ&#41; is another method for distance-based phylogenetics. It is also a clustering method, but one that does not produce ultrametric trees. To run tree inference with NJ for an input alignment or distance matrix, run <code>fastme</code> with the <code>-o &lt;output_file&gt;</code> and <code>-m NJ</code> options. For instance</p>
<pre><code class="language-julia">fastme -i 18SrRNA_20.phy -o 18SrRNA_20_JC.nwk -dJC69 -m NJ</code></pre>
<p>Open the NJ tree in the <code>.nwk</code> file generated by FastME in FigTree &#40;should be straightforward&#41;.</p>
<blockquote>
<p><strong>Question</strong>: How can you see at one glance that this is not an ultrametric tree? What does this mean in terms of assumptions on the substitution rate?</p>
</blockquote>
<blockquote>
<p><strong>Question</strong>: Neighbor-joining infers an unrooted tree, can you see how the fact that the tree is unrooted is represented in FigTree? Where should you root this tree &#40;based on your knowledge of the tree of life&#41;? Select the branch where you think the root should be &#40;click on it&#41; and hit the <code>reroot</code> button in figtree to root the tree.</p>
</blockquote>
<blockquote>
<p><strong>Question</strong>: Does the NJ tree make more sense than the WPGMA tree? What do you think is causing this?</p>
</blockquote>
<p>Now infer a tree using \(\Gamma\) distances</p>
<pre><code class="language-julia">fastme -i 18SrRNA_20.phy -o 18SrRNA_20_JC.nwk -dJC69 -g -m NJ</code></pre>
<blockquote>
<p><strong>Question</strong>: What &#40;if anything&#41; is changing? Experiment with the \(\alpha\) parameter of the Gamma distribution by using for instance <code>-g0.5</code> in the FastME command. What happens? How does the inference for different values of \(\alpha\) relate to the inference without Gamma distances? &#40;FYI: below you can see a graph of the Gamma distribution for different values of \(\alpha\) to help you interpret the results.&#41;</p>
</blockquote>
<pre><code class="language-julia">using Distributions
p &#61; plot&#40;title&#61;&quot;The Gamma distribution with mean 1&quot;&#41;
for α in &#91;0.1, 0.25, 0.5, 1.0, 5.0, 10., 100.&#93;
    plot&#33;&#40;p, Gamma&#40;α, 1/α&#41;, label&#61;&quot;\\alpha &#61; &#36;α&quot;, xlim&#61;&#40;0,5&#41;, ylim&#61;&#40;0,5&#41;&#41;
end</code></pre>
<p><img src="/assets/phylocourse/distance/gamma.svg" alt="" /></p>
<blockquote>
<p><strong>Exercise</strong>: Perform phylogenetic analysis using distance based methods for the second data set with more species. Perform an analysis using the JC model, and the JC&#43;Γ model, and vary the shape of the Gamma distribution in the latter case. Record the differences in the phylogenetic trees you inferred and note them down &#40;we will compare them later with maximum-likelihood results&#41;.</p>
</blockquote>
<blockquote>
<p><strong>Optional</strong>: Explore some of the other substitution models available in FastME.</p>
</blockquote>
<h2 id="extra_implementing_neighbor-joining"><a href="#extra_implementing_neighbor-joining" class="header-anchor">Extra: implementing Neighbor-Joining</a></h2>
<p>Implementing the neighbor-joining algorithm is fairly easy &#40;at least if we don&#39;t care <em>too</em> much about efficiency&#41;. The code below is a fairly minimal implementation of the NJ algorithm &#40;generating the tree directly in Newick format on the go&#41;:</p>
<pre><code class="language-julia">function neighbor_joining&#40;matrix, taxa&#41;
    clades &#61; copy&#40;taxa&#41;
    nodes &#61; collect&#40;1:length&#40;taxa&#41;&#41;
    n &#61; length&#40;nodes&#41;
    while length&#40;nodes&#41; &gt; 1
        # get the next neighbors to join
        &#40;i, j, a, b, di, dj&#41;, new_dist &#61; get_neighbors_to_join&#40;matrix, nodes&#41;
        # join the chosen nodes to  new clade
        push&#33;&#40;clades, &quot;&#40;&#36;&#40;clades&#91;i&#93;&#41;:&#36;di,&#36;&#40;clades&#91;j&#93;&#41;:&#36;dj&#41;&quot;&#41;
        # update the nodes that are still left to join
        nodes&#91;a&#93; &#61; n &#43; 1
        deleteat&#33;&#40;nodes, b&#41;
        # update the matrix with the new node
        matrix &#61; &#91;&#91;matrix ; new_dist&#91;1:end-1&#93;&#39;&#93; new_dist&#93;
        # increment internal node counter
        n &#43;&#61; 1
    end
    clades&#91;end&#93;
end

function get_neighbors_to_join&#40;matrix, nodes&#41;
    r &#61; length&#40;nodes&#41;
    minindex &#61; &#40;0, 0, 0, 0, Inf&#41;
    for a&#61;1:r, b&#61;a&#43;1:r
        i &#61; nodes&#91;a&#93;
        j &#61; nodes&#91;b&#93;
        # This is the neighbor joining optimality criterion, the two nodes that
        # lead to the lowest value of &#96;x&#96; below are chosen in this iteration
        # of the NJ algorithm to join and for an internal node of the tree.
        x &#61; &#40;r-2&#41;*matrix&#91;i,j&#93; - sum&#40;&#91;matrix&#91;i,k&#93; &#43; matrix&#91;j,k&#93; for k in nodes&#93;&#41;
        minindex &#61; x &lt; minindex&#91;end&#93; ? &#40;i, j, a, b, x&#41; : minindex
    end
    i &#61; minindex&#91;1&#93;
    j &#61; minindex&#91;2&#93;
    di, dj, new_distances &#61; get_nj_distance&#40;matrix, nodes, i, j, r&#41;
    return &#40;i, j, minindex&#91;3&#93;, minindex&#91;4&#93;, di, dj&#41;, new_distances
end

function get_nj_distance&#40;matrix, nodes, i, j, r&#41;
    # This is the formula to compute the branch
    a &#61; sum&#40;&#91;matrix&#91;i,k&#93; for k in nodes&#93;&#41;
    b &#61; sum&#40;&#91;matrix&#91;j,k&#93; for k in nodes&#93;&#41;
    if r &#33;&#61; 2  # we are not joining the two last nodes &#40;root&#41;
        di &#61; 0.5*matrix&#91;i,j&#93; &#43; 1.0/&#40;2r - 4&#41;*&#40;a - b&#41;
        dj &#61; 0.5*matrix&#91;i,j&#93; &#43; 1.0/&#40;2r - 4&#41;*&#40;b - a&#41;
    else  # we are joining the two last nodes &#40;generating the root&#41;
        di &#61; matrix&#91;i,j&#93;
        dj &#61; 0.
    end
    new_distances &#61; 0.5 .* &#40;matrix&#91;i,:&#93; .- di .&#43; matrix&#91;j,:&#93; .- dj&#41;
    return di, dj, &#91;new_distances ; 0.&#93;
end</code></pre><pre><code class="plaintext">get_nj_distance (generic function with 1 method)</code></pre>
<p>Then, given a distance matrix, we can use the code like this</p>
<pre><code class="language-julia">matrix, taxa, ntaxa &#61; readmatrix&#40;&quot;_assets/phylocourse/distance/18SrRNA_20_JCGamma_matrix.txt&quot;&#41;
neighbor_joining&#40;matrix, taxa&#41;</code></pre><pre><code class="plaintext">"((((Homosapie:0.02924755103488889,Xenopusla:0.021778841927111107):0.08114770060050783,Artemiasa:0.11616511535899218):0.06494892127699999,Anemonias:0.07924348839900001):0.009972187360844695,((((((Oryzasati:0.0061716773969643,Zeamaysxx:0.0143528379960357):0.02197552496695191,Glycinema:0.025865134953548086):0.01417354623424481,Zamiapumi:0.041985046430755193):0.03377916779845535,((Chlamydomo:0.010749288627016637,Volvoxcar:0.009419747100983361):0.031210936093406267,Chlorellaa:0.04110170407209374):0.03217114798238841):0.02786998226327543,Palmariap:0.2015992260026777):0.01282991248020214,(((Candidaal:0.019243759892322902,(Kluyveromy:0.008400335147631251,Saccharomy:0.013484863547368749):0.018706281291177095):0.035214710391360926,Neurospora:0.07166376672263906):0.011126122879444332,((Bulleraal:0.020002847598359373,Trichospor:0.022424966190640626):0.027329815263197922,(Leucospori:0.0098172971579853,Sporobolom:0.011305443089014702):0.08077801428405207):0.012930904895555671):0.030469872136733397):0.0)"</code></pre>
<p>You can check this against FastME&#39;s NJ implementation, it should be correct. We have everything in place to start building our own extremely redundant phylogenetics library.</p>
<blockquote>
<p><strong>Exercise</strong>: For the diehards, try to understand the code and perhaps reimplement it in your programming language of choice.</p>
</blockquote>
<hr />
<p><table class="fndef" id="fndef:fastmeonline">
    <tr>
        <td class="fndef-backref"><a href="#fnref:fastmeonline">[1]</a></td>
        <td class="fndef-content">Note that you can also run FastME online <a href="http://www.atgc-montpellier.fr/fastme/">here</a>.</td>
    </tr>
</table>
 <table class="fndef" id="fndef:ancientdata">
    <tr>
        <td class="fndef-backref"><a href="#fnref:ancientdata">[2]</a></td>
        <td class="fndef-content">These data sets are from rather ancient days in phylogenetics, and are stored in the so-called PHYLIP format. In the original <a href="http://evolution.genetics.washington.edu/phylip/doc/main.html#inputfiles">PHYLIP format</a>, taxon names were restricted to 10 characters, hence the weird truncated names for the sequences.</td>
    </tr>
</table>
 <table class="fndef" id="fndef:commandline">
    <tr>
        <td class="fndef-backref"><a href="#fnref:commandline">[3]</a></td>
        <td class="fndef-content">For those unfamiliar with the command line, it will probably be easiest to put the FastME executable for your operating system &#40;for windows this is the <code>fastme.exe</code> file&#41; together with the data files in a separate new folder. On windows, you can then from within your file explorer application do <code>Shift&#43;right click</code> and click on <code>open command window here</code> or <code>open PowerShell window here</code>. Then you should be able to run the <code>fastme.exe</code> executable, e.g. <code>fastme -i 18SrRNA_20.phy -O 18SrRNA_20_JCmatrix.txt -d JC</code> provided the data file <code>18SrRNA_20.phy</code> is in the same directory as the <code>fastme.exe</code> executable. For Mac and linux users I would recommend a similar approach, make a directory where you put the executable and the data files, and open a terminal in that directory &#40;Linux users, you know how to do this, MacOS users, I can&#39;t help you, but google is your friend&#41;. See also some notes <a href="/phylocourse/install-windows">here</a></td>
    </tr>
</table>
</p>

<div class="page-foot">
  <div class="copyright">
    &copy; Arthur Zwaenepoel. Last modified: March 31, 2021. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>
</div><!-- CONTENT ENDS HERE -->
    
        <script src="/libs/katex/katex.min.js"></script>
<script src="/libs/katex/auto-render.min.js"></script>
<script>renderMathInElement(document.body)</script>

    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
  </body>
</html>
