<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <!-- Begin Jekyll SEO tag v2.4.0 -->
<title>Recap - finite populations | bb</title>
<meta name="generator" content="Jekyll v3.7.3" />
<meta property="og:title" content="Recap - finite populations" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Finite idealized population" />
<meta property="og:description" content="Finite idealized population" />
<link rel="canonical" href="http://localhost:4000/qgen%20popgen%20evolution%20recap/2018/04/30/finite_populations.html" />
<meta property="og:url" content="http://localhost:4000/qgen%20popgen%20evolution%20recap/2018/04/30/finite_populations.html" />
<meta property="og:site_name" content="bb" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2018-04-30T00:00:00+02:00" />
<script type="application/ld+json">
{"description":"Finite idealized population","@type":"BlogPosting","url":"http://localhost:4000/qgen%20popgen%20evolution%20recap/2018/04/30/finite_populations.html","headline":"Recap - finite populations","dateModified":"2018-04-30T00:00:00+02:00","datePublished":"2018-04-30T00:00:00+02:00","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/qgen%20popgen%20evolution%20recap/2018/04/30/finite_populations.html"},"@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

  <link rel="stylesheet" href="/assets/main.css">
  <link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="bb" />
  
</head>


  <body>

    <header class="site-header" role="banner">

  <div class="wrapper">
    
    
    <a class="site-title" rel="author" href="/">bb</a>

    
      <nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
              <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
              <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger">
          
            
            
          
            
            
            <a class="page-link" href="/about/">About</a>
            
          
            
            
          
            
            
          
            
            
          
        </div>
      </nav>
    
  </div>
</header>


    <main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Recap - finite populations</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2018-04-30T00:00:00+02:00" itemprop="datePublished">
        
        Apr 30, 2018
      </time>
      </p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="finite-idealized-population">Finite idealized population</h2>

<p>A bit of a recap is never a bad thing to do. So considering a locus with two alleles $p$ and $q$ in a finite diploid population of size $N$, initially at frequencies $q_0$ and $p_0$. Assuming an idealized population, following all Hardy-Weinberg conditions except the infinite population size, the expected frequency of $q_1$ is then $q_0$. The variance of this quantity can be calculated as a binomial sampling variance with sample size of $2N$ and is given by:</p>

<script type="math/tex; mode=display">\sigma^2_{q_1} = \frac{q_0 (1-q_0)}{2N}</script>

<p>This can be interpreted as the variance in $q_1$ among different samples of size $N$ drawn from a base population with $q = q_0$. This is also the variance of the <em>change</em> in allele frequency $\sigma^2_{\Delta q}$.</p>

<p>We can also consider <strong>genotype frequencies</strong>, where the frequency of homozygotes in generation $t$ is of course given by $q_{t-1}^2$ and $p_{t-1}^2$. Consider the mean frequency of homozygotes among lines sampled from a base population with minor allele frequency $q_0$, which can be denoted $\overline{q^2}$. Recall that the expression for the variance of a random variable $X$ is given by $E(X^2) - E(X)^2$. Thus we have:</p>

<script type="math/tex; mode=display">% <![CDATA[
\begin{eqnarray}
\sigma^2_{q} &=& \overline{q^2} - \overline{q}^2 \\
\overline{q^2} &=& \overline{q}^2 + \sigma^2_{q}
\end{eqnarray} %]]></script>

<p>The consequences of this expression are clear. Indeed, $\overline{q}$ is just given by $q_0$, <em>i.e.</em> the allele frequency in the base population. So the expected number of homozygotes $\overline{q^2}$ is always in excess over the original fraction of homozygotes $q_0^2$. Likewise, the fraction of heterozygotes is reduced by $2\sigma^2_{q}$. Note that we have not considered the process in function of time. We have just observed a snapshot situation in a finite population and have related it to the base population from which it was originally drawn. By doing this we observed that the fraction of homozygotes is increased with the variance of the allele frequency $\sigma_q^2$. How $\sigma_q^2$ evolves with time is something we will consider next.</p>

<h2 id="inbreeding">Inbreeding</h2>

<p>We will now consider the famous coefficient of inbreeding, which is a measure of the dispersive process happening in finite populations. The coefficient of inbreeding $F$ is the probability that two alleles at a randomly sampled locus are <em>identical by descent</em> (IBD). The average $F$ gives a measure of genetic drift in a population. Note that $F$ is only meaningful when compared to a base population!</p>

<p>Considering a base population with $2N$ genes which we consider all to be independent. We can easily see that $ F_1 = \frac{1}{2N} $. Similarly, we can reason that</p>

<script type="math/tex; mode=display">F_2 = \frac{1}{2N} + (1 - \frac{1}{2N})F_1</script>

<p>This gives a recursion for $F$. However this recursion is as such not very easy to work with. We can easily see that there is always an addition of $\frac{1}{2N}$ to the inbreeding coefficient in subsequent generations, originating from the probability to sample identical alleles in the base population. Call this $\Delta F = \frac{1}{2N}$. Now we can write the following expression</p>

<script type="math/tex; mode=display">F_{t} = \Delta F + (1 -\Delta F)F_{t-1}</script>

<p>which can be rearranged to give</p>

<script type="math/tex; mode=display">\Delta F = \frac{F_{t} - F_{t-1}}{1-F_{t-1}}</script>

<p>which clearly shows the nature of $\Delta F$ as a measure for the rate of inbreeding. Indeed, it expresses the increase in the inbreeding coefficient normalized by the proportion of the population that was not yet inbred.</p>

<p>We now define the panmictic index $P_t = 1-F_t$ and we rearrange the last expression to give</p>

<script type="math/tex; mode=display">\frac{P_t}{P_{t-1}} = 1-\Delta F</script>

<p>which gives rise to a lovely recursion rendering</p>

<script type="math/tex; mode=display">% <![CDATA[
\begin{eqnarray}
\frac{P_t}{P_{t-2}} = \frac{P_t}{P_{t-1}} \frac{P_{t-1}}{P_{t-2}} &=& (1-\Delta F)^2 \\
&\vdots& \\
P_t &=& (1 - \Delta F)^t
\end{eqnarray} %]]></script>

<p>Which allows us to get a general expression for $F$ in function of $t$</p>

<script type="math/tex; mode=display">F_t = 1 - (1-\Delta F)^t</script>

<p>which holds of course only when the F_0 is considered as 0.</p>

<p>We can now revisit the expressions in the previous sections which were attained by considering properties of binomial samples. We have shown that $\sigma_{\Delta q}^2 = \frac{q_0(1-q_0)}{2N}$ which we can now express as $q_0(1-q_0)\Delta F = q_0p_0\Delta F$. Similarly we can consider the variance in gene frequencies among different hypothetical lines of a same base population in any generation as $\sigma_{q,t}^2 = q_0p_0F_t$. Whereas $\Delta F$ can be regarded as the rate of new inbreeding, $F_t$ is a measure of the accumulated effect of random drift over $t$ generations.</p>

<p>We can again consider genotype frequencies. The mean fraction of homozygotes across hypothetical lines of size $2N$ of a same base population is still given by</p>

<script type="math/tex; mode=display">\overline{q^2} = \overline{q}^2 + \sigma_q^2</script>

<p>Where $\sigma_q^2$ can now be expressed in terms of the base population frequencies and the number of generations! Indeed, we have:</p>

<script type="math/tex; mode=display">\overline{q^2} = q_0^2 + q_0p_0F_t</script>

<p>This clearly shows the cumulating excess of homozygotes through inbreeding. Similarly, we can consider the decline in the fraction heterozygotes ($H$). Given $H_0 = 2p_0q_0$ we can see that</p>

<script type="math/tex; mode=display">% <![CDATA[
\begin{eqnarray}
H_t &=& 2p_0q_0 - 2q_0p_0F_t \\
    &=& 2p_0q_0(1-F_t) \\
    &=& H_0(1-F_t) = H_0P_t
\end{eqnarray} %]]></script>

<p>since the homozygote excess occurs at the expense of heterozygotes. Interestingly, this provides an interpretation for the panmictic index as the proportional decrease in the fraction of heterozygotes relative to the base population.</p>

<p>This is I think the core theory of inbreeding and basic population genetics in finite populations which one should always be able to reproduce on demand.</p>

  </div>

  

  <a class="u-url" href="/qgen%20popgen%20evolution%20recap/2018/04/30/finite_populations.html" hidden></a>
</article>

<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript" ></script>
<script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: { autoNumber: "AMS" } } }); </script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">bb</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">
            
              bb
            
            </li>
            
            <li><a class="u-email" href="mailto:arzwa@psb.ugent.be">arzwa@psb.ugent.be</a></li>
            
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
  
  
  
  <li><a href="https://github.com/arzwa"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">arzwa</span></a></li>
  
  
  
  
  
  
  
  <li><a href="/feed.xml"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#rss"></use></svg> <span>rss</span></a></li>
</ul>

      </div>

      <div class="footer-col footer-col-3">
        <p>Take a moment to describe yourself</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
