<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  
   <link rel="stylesheet" href="/libs/highlight/github.min.css">
   
  <link rel="stylesheet" href="/css/franklin.css">
  <link rel="stylesheet" href="/css/basic.css">
   <title>Bayesian phylogenetic inference</title>  
</head>
<body>
  <header>
	  <!--div class="blog-name"><a href="/">bb</a></div--->
  </header>

<!-- Content appended here -->
<div class="franklin-content"><p><a href="/phylocourse/">back</a> <div class="franklin-toc"><ol><li><a href="#bayesian_phylogenetic_inference">Bayesian phylogenetic inference</a></li><li><a href="#mrbayes_exercise">MrBayes exercise</a></li></ol></div></p>
<h2 id="bayesian_phylogenetic_inference"><a href="#bayesian_phylogenetic_inference" class="header-anchor">Bayesian phylogenetic inference</a></h2>
<p>A theoretical introduction will be given in-class.</p>
<h2 id="mrbayes_exercise"><a href="#mrbayes_exercise" class="header-anchor">MrBayes exercise</a></h2>
<p>Download and install <a href="https://github.com/NBISweden/MrBayes/releases/tag/v3.2.7">MrBayes</a>. Some more information can be found <a href="http://nbisweden.github.io/MrBayes/download.html">here</a>. Windows users should use the precompiled executables in the <code>*-WIN.zip</code> file. Macos users should be able to install it via Homebrew. Linux users should have no difficulty building from source as indicated in the <a href="https://github.com/NBISweden/MrBayes">README</a>.</p>
<p>The <a href="https://github.com/NBISweden/MrBayes/blob/develop/doc/manual/Manual_MrBayes_v3.2.pdf">manual for MrBayes</a> provides very helpful information.</p>
<p>You will need a file in the <code>nexus</code> format to use MrBayes. Tip, if you have python installed, then the Biopython library provides convenient alignment format conversion tools:</p>
<pre><code class="language-julia">from Bio import AlignIO
from Bio.Alphabet import DNAAlphabet
AlignIO.convert&#40;&quot;inputfile.fasta&quot;, &quot;fasta&quot;, &quot;outputfile.nex&quot;, &quot;nexus&quot;, alphabet&#61;DNAAlphabet&#40;&#41;&#41;</code></pre>
<p>Anyway, <a href="/assets/phylocourse/data/hiv-dentist.nex">here</a> is a nexus file for the HIV-dentist data set.</p>
<p>Open up the MrBayes prompt by typing <code>mb</code> at the command line. You should see something like the following</p>
<pre><code class="language-julia">&#36; mb


                           MrBayes 3.2.7a x86_64

                     &#40;Bayesian Analysis of Phylogeny&#41;

             Distributed under the GNU General Public License


              Type &quot;help&quot; or &quot;help &lt;command&gt;&quot; for information
                    on the commands that are available.

                  Type &quot;about&quot; for authorship and general
                      information about the program.


MrBayes &gt;</code></pre>
<p>Now load the data file by typing <code>execute hiv-dentist.nex</code>. You can specify a substitution model using <code>lset</code>. For instance <code>lset nst&#61;2 rates&#61;gamma
ngammacat&#61;4</code> this will set the nucleotide substitution model &#40;<code>nst</code>&#41; to model 2 &#40;which is the K2P model&#41;, assume Gamma distributed rates across sites and will approximate the Gamma distribution by four discrete rate categories.</p>
<blockquote>
<p><strong>Question:</strong> To do Bayesian inference we will need to make prior assumptions under the form of prior distributions for the parameters of our model. In the basic model as we have set it up here &#40;<code>K2P&#43;G4</code>&#41; which model parameters are there? Type <code>help prset</code> and have a look at the default prior settings that are currently selected. Most of the prior settings listed here are not relevant for the current model, which prior settings are relevant for the currently specified model? You can check this by typing <code>showmodel</code>.</p>
</blockquote>
<p>The default prior settings in MrBayes are well-thought of and designed to do a good job in many cases. They are more or less so-called &#39;uninformative&#39; or &#39;vague&#39; priors, so that the results are not, in general, very sensitive to them. Nevertheless, this <strong>cannot be an excuse</strong> to not be aware of the prior settings and parameters&#33; If one does Bayesian inference, one does commit oneself to the full deal.</p>
<p>As an example of how we may wish to adjust the priors, consider the branch lengths prior.</p>
<blockquote>
<p><strong>Question:</strong> What is the default branch length prior distribution? Read about it in the documentation available in <code>help prset</code>. Do you understand how this how this prior is defined? Change the prior to something more simple, such as an exponential distribution for each branch. Choose the parameter of the exponential prior based on your Maximum likelihood trees inferred with IQ-TREE.</p>
</blockquote>
<p>Now let&#39;s run the MCMC algorithm. Type <code>mcmc ngen&#61;100000 samplefreq&#61;100
printfreq&#61;1000 diagnfreq&#61;10000</code>.</p>
<blockquote>
<p><strong>Question:</strong> By default MrBayes will run two MCMC chains, why do we run a second chain, what could it tell us that a single chain can&#39;t? What does the standard deviation of split frequencies mean?</p>
</blockquote>
<p>If the standard deviation of split frequencies is low enough &#40;below 0.01, or below 0.05 if you&#39;re inpatient&#41; you can stop the run. After stopping the run you can type <code>sump</code> to summarize the posterior distribution for model parameters and <code>sumt</code> to summarize the posterior distribution for the trees.</p>
<blockquote>
<p><strong>Question:</strong> Interpret the output of <code>sumt</code> printed to the screen, what would be the posterior mean estimate for the transition/transversion ratio &#40;<code>kappa</code>&#41;? What is its 95&#37; uncertainty interval &#40;and what does that mean?&#41;. What is the ESS? Have we run the chain long enough?</p>
</blockquote>
<blockquote>
<p><strong>Question:</strong> Now look at the <code>sumt</code> output, what do the clade credibility values mean?</p>
</blockquote>
<p>To visually explore the samples from the posterior distributions, you can use the programs <a href="https://github.com/beast-dev/tracer/releases/tag/v1.6"><code>Tracer</code></a> and <a href="https://github.com/rambaut/figtree/releases/tag/v1.4.4"><code>FigTree</code></a>. You can load the <code>*.t</code> output file&#40;s&#41; in FigTree and browse through the sampled trees. You can load the <code>*.p</code> output file&#40;s&#41; in Tracer to check the posterior distribution &#40;you may need to delete the first line from the <code>*.p</code> output file from MrBayes to load it in Tracer&#41;.</p>
<blockquote>
<p><strong>Question:</strong> Look at the trace plots for the different parameters. Did we manage to reach the stationary distribution of the MCMC chain?</p>
</blockquote>

<div class="page-foot">
  <div class="copyright">
    &copy; Arthur Zwaenepoel. Last modified: January 10, 2022. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>.
  </div>
</div>
</div><!-- CONTENT ENDS HERE -->
    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
  </body>
</html>
